---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
library(dplyr)
devtools::load_all()
```
[![Build Status](https://travis-ci.org/sa-lee/plyranges.svg?branch=master)](https://travis-ci.org/sa-lee/plyranges)

# plyranges
The `plyranges` ^[tentative title] package provides a 
[fluent](https://en.wikipedia.org/wiki/Fluent_interface) interface for 
performing common genomic data analysis tasks. One goal of this package is to 
decrease the learning curve for Bioconductor and S4 classes (especially for new users) by providing a consistent interface to common classes, _IRanges_, _GRanges_ and _SummarizedExperiment_. 

To achieve this we have constructed a grammar of genomic data manipulation
based on `dplyr` and the Bioconductor packages that define the aformentioned
classes. All the methods defined in `plyranges` are type-consistent and 
chainable. As a consequence, users familiar with 
`dplyr` should be able to read and understand `plyranges` code without 
difficulty.

*Note this package is under heavy development*

# Installation

Currently only the development version is available
```{r, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("sa-lee/plyranges")
```

# Constructing Ranges

Ranges can be easily constructed from a data.frame or tibble. To construct an
`IRanges` we require there are at least two columns that represent at either a 
starting coordinate, finishing coordinate or the width of the interval. To
construct a `GRanges` we require a column that represents that sequence name (
contig or chromosome id), and an optional column to represent the strandedness
of an interval. 

```{r}
set.seed(100)
df <- data.frame(start=c(2:-1, 13:15), 
                 width=c(0:3, 2:0))

# produces IRanges
df %>% Ranges()

# produces GRanges
df %>% 
  mutate(seqnames = sample(c("chr1", "chr2"), 7, replace = TRUE)) %>% 
  Ranges()

df  %>% 
  mutate(strand = sample(c("+", "-"), 7, replace = TRUE)) %>% 
  Ranges()

# seqname is required for GRanges, metadata is automatically kept
df %>% 
  mutate(seqnames = sample(c("chr1", "chr2"), 7, replace = TRUE),
         strand = sample(c("+", "-"), 7, replace = TRUE),
         gc = runif(7)) %>% 
  Ranges()

```

### Modifying Genomic Intervals

Sometimes you want to modify a genomic interval by altering it's coordinateswidth and
leave the start, end or midpoint of the coordinates unaltered. This is achieved
with the `set_width` verb along with `anchor_*` adverbs.
```{r}
rng <- Ranges(df)
set_width(rng, 100L)
set_width(anchor_start(rng), 100L)
set_width(anchor_end(rng), 100L)
set_width(anchor_center(rng), 100L)
```

Similarly, you can modify the width of an interval using the `stretch` verb.
Without anchoring, this function will extend the interval in either direction
by an integer amount. With anchoring, either the start, end or midpoint are
preserved. 

```{r}
rng2 <- stretch(anchor_center(rng), 10)
rng2
stretch(anchor_end(rng2), 1L)
stretch(anchor_start(rng2), 10L)
```

_Ranges_ can be shifted left or right.

```{r}
shift_left(rng, 10L)
shift_right(rng, 10L)
```


