---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
library(dplyr)
devtools::load_all()
```

# plyranges

The `plyranges` ^[tentative title] package provides a 
[fluent](https://en.wikipedia.org/wiki/Fluent_interface) interface for 
performing common genomic data analysis tasks. One goal of this package is to 
decrease the learning curve for Bioconductor and S4 classes (especially for new users) by providing a consistent interface to common classes, _IRanges_, _GRanges_ and _SummarizedExperiment_. 

To achieve this we have constructed a grammar of genomic data manipulation
based on `dplyr` and the Bioconductor packages that define the aformentioned
classes. All the methods defined in `plyranges` are type-consistent and 
chainable. As a consequence, users familiar with 
`dplyr` should be able to read and understand `plyranges` code without 
difficulty.

*Note this package is under heavy development*

# Installation

Currently only the development version is available
```{r, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("sa-lee/plyranges")
```

# Constructing Ranges

Ranges can be easily constructed from a data.frame or tibble. To construct an
`IRanges` we require there are at least two columns that represent at either a 
starting coordinate, finishing coordinate or the width of the interval. To
construct a `GRanges` we require a column that represents that sequence name (
contig or chromosome id), and an optional column to represent the strandedness
of an interval. 

```{r}
library(dplyr)
df <- data.frame(start=c(2:-1, 13:15), 
                 width=c(0:3, 2:0))

df %>% Ranges()

df %>% 
  mutate(seqnames = sample(c("chr1", "chr2"), nrow(df), replace = TRUE),
         strand = "*") %>% 
  Ranges()
```


