---
title: "Getting started with the plyranges package"
author: "Stuart Lee"
package: plyranges
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

# `Ranges` revisted

In Bioconductor there are two classes, `IRanges` and `GRanges`, 
that are standard data structures for representing genomics data. 
Throughout this document I refer to either of these classes as `Ranges` if an
operation can be performed on either class, otherwise I explicilty mention if
a function is approrpriate for an `IRanges` or `GRanges`.

`Ranges` objects can either represent sets of integers as `IRanges`
(which have start, end and width attributes)
or represent genomic intervals (which have additional attributes, sequence name,
and strand) as `GRanges`. In addition, both types of `Ranges` can store information
about their intervals as metadata columns (for example GC content 
over a genomic interval). 

`Ranges` objects follow the tidy data principle: each row
of a `Ranges` object corresponds to an interval, while each column will represent
a variable about that interval, and generally each object will represent a
single unit of observation (like gene annotations). 

Consequently, `Ranges` objects provide a powerful representation for reasoning
about genomic data. In this vignette, you will learn more about `Ranges` objects
and how via grouping, restriction and summarisation you can perform common data
tasks. 


# Constructing `Ranges`

To construct an `IRanges` we require that there are at least two columns that represent at either a 
starting coordinate, finishing coordinate or the width of the interval. 

```{r}
suppressPackageStartupMessages(library(plyranges))
set.seed(100)
df <- data.frame(start=c(2:-1, 13:15), 
                 width=c(0:3, 2:0))

# produces IRanges
rng <- df %>% as_iranges()
rng

```

To construct a `GRanges` we require a column that represents that sequence name (
contig or chromosome id), and an optional column to represent the strandedness
of an interval. 

```{r}
# seqname is required for GRanges, metadata is automatically kept
grng <- df %>% 
  transform(seqnames = sample(c("chr1", "chr2"), 7, replace = TRUE),
         strand = sample(c("+", "-"), 7, replace = TRUE),
         gc = runif(7)) %>% 
  as_granges()

grng
```

# Arithmetic on Ranges

Sometimes you want to modify a genomic interval by altering the width of the 
interval while leaving the start, end or midpoint of the coordinates unaltered. 
This is achieved with the `set_width` verb along with `anchor_*` adverbs.
```{r}
rng <- as_iranges(data.frame(start=c(1, 2, 3), end=c(5, 2, 8)))
grng <- as_granges(data.frame(start=c(1, 2, 3), end=c(5, 2, 8), 
                          seqnames = "seq1",
                          strand = c("+", "*", "-")))
set_width(rng, 10)
set_width(anchor_start(rng), 10)
set_width(anchor_end(rng), 10)
set_width(anchor_center(rng), 10)
set_width(anchor_3p(grng), 10) # leave negative strand fixed
set_width(anchor_5p(grng), 10) # leave positve strand fixed
```

Similarly, you can modify the width of an interval using the `stretch` verb.
Without anchoring, this function will extend the interval in either direction
by an integer amount. With anchoring, either the start, end or midpoint are
preserved. 

```{r}
rng2 <- stretch(anchor_center(rng), 10)
rng2
stretch(anchor_end(rng2), 10)
stretch(anchor_start(rng2), 10)
stretch(anchor_3p(grng), 10)
stretch(anchor_5p(grng), 10)
```

`*Ranges` can be shifted left or right. If strand information is available 
we can also shift upstream or downstream. 

```{r}
shift_left(rng, 10)
shift_right(rng, 10)
shift_upstream(grng, 10)
shift_downstream(grng, 10)
```

# Grouping `Ranges`

`plyranges` introduces a new class of `Ranges` called `RangesGrouped`, 
this is a similiar idea to the grouped `data.frame\tibble` in `dplyr`. 

Grouping can act on either the core components or the metadata columns of
a `Ranges` object.

It is most effective when combined with other verbs such as `mutate()`,
`summarise()`, `filter()`, `reduce_ranges()` or `disjoin_ranges()`.
```{r}
grng <- data.frame(seqnames = sample(c("chr1", "chr2"), 7, replace = TRUE),
         strand = sample(c("+", "-"), 7, replace = TRUE),
         gc = runif(7),
         start = 1:7,
         width = 10) %>%
  as_granges()

grng_by_strand <- grng %>%
  group_by(strand)

grng_by_strand
```

# Restricting `Ranges`

The verb `filter` can be used to restrict rows in the `Ranges`.
Note that grouping will cause the `filter` to act within each group 
of the data.
```{r}
grng %>% filter(gc < 0.3)
# filtering by group 
grng_by_strand %>% filter(gc == max(gc)) 
```

We also provide the convience methods `filter_by_overlaps` and 
`filter_by_non_overlaps` for restricting by overlapping `Ranges`.

```{r}
ir0 <- data.frame(start = c(5,10, 15,20), width = 5) %>%
  as_iranges()
ir1 <- data.frame(start = 2:6, width = 3:7) %>%
  as_iranges()
ir0
ir1
ir0 %>% filter_by_overlaps(ir1)
ir0 %>% filter_by_non_overlaps(ir1) 
```

# Summarising `Ranges`

The `summarise` function will return a tibble because the information required
to return a `Ranges` object is lost. It is often most useful to use `summarise`
in combination with the `group_by` family of functions. 

```{r}
ir1 <- ir1 %>%
  mutate(gc = runif(length(.)))

ir0 %>% 
  group_by_overlaps(ir1) %>%
  summarise(gc = mean(gc))
```

# Joins, or another way at looking at overlaps between `Ranges`

![Types of overlap](overlaps.png)

We can think about finding overlaps as a type of join between two `Ranges`
where we are using the intervals as keys. Of most use are the 
`join_overlap_intersect`, `join_overlap_inner` and `join_overlap_left` functions. 

The first function, `join_overlap_intersect` will return a `Ranges`` object 
where the start, end, and width coordinates correspond to the amount of any overlap
between the left and right input `Ranges`.

![intersect join](intersects.png)


These coordinates correspond to the intervals that are the intersecting `Ranges`. 
By default, the input `Ranges` (and any other metadata) will be carried over as 
metadata columns. The names of the columsn

```{r}
join_overlap_intersect(ir0, ir1, suffix = c(".left", ".right"))
```

The `join_overlap_inner` function will return the `Ranges` on the left that
overlap any `Ranges` on the right. By default, it returns the `Ranges` on the right
that overlap (and any other metadata columns belonging to the left or right)
as a metadata column.

![inner join](inner_join.png)


```{r}
join_overlap_inner(ir0, ir1, suffix = c(".left", ".right"))
```

We also provide a convienence method called `find_overlaps` that computes the 
same result as `join_overlap_inner`.

```{r}
find_overlaps(ir0, ir1, suffix = c(".left", ".right"))
```

The `join_overlap_left` method will perform an outer left join. 

![left join](left_join.png)

First any overlaps that are found will be returned similar to `join_overlap_inner`. 
Then any non-overlapping ranges will be returned, with missing values on the metadata columns. 

A 'missing' value for an `IRanges` column is interval with zero width,
with `start=0` and `end=-1`. A 'missing' value for a `GRanges` column is the same
as the `IRanges` but with `seqnames='.'` and `strand='*'`.

```{r}
join_overlap_left(ir0, ir1)
```

Compared with `filter_by_overlaps` above, the overlap left join expands the
`Ranges` to give information about each interval on the left hand side that 
overlap those on the right as well as the intervals on the left that do not overlap any range on the right.

## Example: dealing with multimapping
This example is taken from the Bioconductor support [site](https://support.bioconductor.org/p/100046/).

We have two `Ranges` objects. The first contains single nucleotide positions
corresponding to an intensity measurement such as a ChiP-seq experiment, 
while the other contains coordinates for two genes of interest.


We want to identify which positions in the `intensties` `Ranges` overlap the
genes, where each row corresponds to a position that overlaps a single gene.

First we create the two `Ranges` objects
```{r ex1}
intensities <- data.frame(seqnames = "VI",
                          start = c(3320:3321,3330:3331,3341:3342),
                          width = 1)
intensities <- as_granges(intensities)

intensities 

genes <- data.frame(seqnames = "VI", 
                    start = c(3322, 3030),
                    end = c(3846, 3338),
                    gene_id=c("YFL064C", "YFL065C"))
                    
                    
genes <- as_granges(genes)
genes

```

Now to find where the positions overlap each gene, we can perform an overlap
join. This will automatically carry over the gene_id information as well as
their coordinates (we can drop those by only selecting the gene_id). 
```{r}
olap <- join_overlap_inner(intensities, genes) %>%
  select(gene_id)
olap
```

Several positions match to both genes. We can count them using `summarise`
and grouping by the `start` position:

```{r}
olap %>% 
  group_by(start) %>%
  summarise(n = n())
```

## Finding your neighbours
We also provide methods for finding nearest, preceding or following `Ranges`.

![Nearest neighbour joins](neighbours.png)

These methods are similar to the overlap joins described above. 

```{r}
join_nearest(ir0, ir1)
join_follow(ir0, ir1)
join_precede(ir0, ir1) # nothing precedes returns empty `Ranges`
join_precede(ir1, ir0)
```

For `GRanges` objects by default strand is not considered when performing
overlap joins. To include strand when finding overlaps use append the prefix
`directed`. To restrict overlapping `Ranges` to those within the query range
use the prefix `within`.

## Grouping by overlaps

It's also possible to group by overlaps. Using this approach we can count the
number of overlaps that are greater than 0. 

```{r}
grp_by_olap <- ir0 %>% 
  group_by_overlaps(ir1)
grp_by_olap
grp_by_olap %>%
  mutate(n_overlaps = n())
```

Of course we can also add overlap counts via the `count_overlaps` function.

```{r}
ir0 %>%
  mutate(n_overlaps = count_overlaps(., ir1))
```

# Data Import/Output

We provide convienence functions via `rtracklayer` 
for reading/writing the following data formats from/to `Ranges` objects.

* BED: `read/write_bed`
* BEDGraph: `read/write_bedgraph`
* narrowPeaks: `read/write_narrowpeaks`
* GFF(1-3): `read/write_gff(1-3)`
* BigWig: `read/write_bw`
* Wig: `read/write_wig`

# Appendix

```{r}
sessionInfo()
```

