<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started with the plyranges package â€¢ plyranges</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with the plyranges package">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">plyranges</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/an-introduction.html">Introduction to plyranges</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/plyranges">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Getting started with the plyranges package</h1>
                        <h4 class="author">Stuart Lee</h4>
            
            <h4 class="date">2018-03-11</h4>
          </div>

    
    
<div class="contents">
<section id="ranges-revisted" class="level1"><h1>
<code>Ranges</code> revisted</h1>
<p>In Bioconductor there are two classes, <code>IRanges</code> and <code>GRanges</code>, that are standard data structures for representing genomics data. Throughout this document I refer to either of these classes as <code>Ranges</code> if an operation can be performed on either class, otherwise I explicilty mention if a function is approrpriate for an <code>IRanges</code> or <code>GRanges</code>.</p>
<p><code>Ranges</code> objects can either represent sets of integers as <code>IRanges</code> (which have start, end and width attributes) or represent genomic intervals (which have additional attributes, sequence name, and strand) as <code>GRanges</code>. In addition, both types of <code>Ranges</code> can store information about their intervals as metadata columns (for example GC content over a genomic interval).</p>
<p><code>Ranges</code> objects follow the tidy data principle: each row of a <code>Ranges</code> object corresponds to an interval, while each column will represent a variable about that interval, and generally each object will represent a single unit of observation (like gene annotations).</p>
<p>Consequently, <code>Ranges</code> objects provide a powerful representation for reasoning about genomic data. In this vignette, you will learn more about <code>Ranges</code> objects and how via grouping, restriction and summarisation you can perform common data tasks.</p>
</section><section id="constructing-ranges" class="level1"><h1>Constructing <code>Ranges</code>
</h1>
<p>To construct an <code>IRanges</code> we require that there are at least two columns that represent at either a starting coordinate, finishing coordinate or the width of the interval.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(plyranges))</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">100</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">2</span><span class="op">:-</span><span class="dv">1</span>, <span class="dv">13</span><span class="op">:</span><span class="dv">15</span>), </a>
<a class="sourceLine" id="cb1-4" data-line-number="4">                 <span class="dt">width=</span><span class="kw">c</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">2</span><span class="op">:</span><span class="dv">0</span>))</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co"># produces IRanges</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">rng &lt;-<span class="st"> </span>df <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>()</a>
<a class="sourceLine" id="cb1-8" data-line-number="8">rng</a></code></pre></div>
<pre><code>## IRanges object with 7 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         1         0
##   [2]         1         1         1
##   [3]         0         1         2
##   [4]        -1         1         3
##   [5]        13        14         2
##   [6]        14        14         1
##   [7]        15        14         0</code></pre>
<p>To construct a <code>GRanges</code> we require a column that represents that sequence name ( contig or chromosome id), and an optional column to represent the strandedness of an interval.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># seqname is required for GRanges, metadata is automatically kept</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">grng &lt;-<span class="st"> </span>df <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="st">  </span><span class="kw">transform</span>(<span class="dt">seqnames =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">         <span class="dt">strand =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">         <span class="dt">gc =</span> <span class="kw">runif</span>(<span class="dv">7</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">grng</a></code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr1  [ 2,  1]      + |  0.76255108229816
##   [2]     chr1  [ 1,  1]      - | 0.669021712383255
##   [3]     chr2  [ 0,  1]      + | 0.204612161964178
##   [4]     chr1  [-1,  1]      - | 0.357524853432551
##   [5]     chr1  [13, 14]      - | 0.359475114848465
##   [6]     chr1  [14, 14]      + | 0.690290528349578
##   [7]     chr2  [15, 14]      + | 0.535811153938994
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</section><section id="arithmetic-on-ranges" class="level1"><h1>Arithmetic on Ranges</h1>
<p>Sometimes you want to modify a genomic interval by altering the width of the interval while leaving the start, end or midpoint of the coordinates unaltered. This is achieved with the <code>set_width</code> verb along with <code>anchor_*</code> adverbs.</p>
<p>The act of anchoring fixes either the start, end, center coordinates of the <code>Range</code> object, as shown in the figure and code below and anchors are used in combination with either <code>set_width</code> or <code>stretch</code>.</p>
<p><img src="anchors.png" width="400px" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>(<span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">8</span>)))</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">grng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>(<span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">8</span>), </a>
<a class="sourceLine" id="cb5-3" data-line-number="3">                          <span class="dt">seqnames =</span> <span class="st">"seq1"</span>,</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">                          <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"-"</span>)))</a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(rng, <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_start</a></span>(rng), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_end</a></span>(rng), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4         5        10
##   [2]        -7         2        10
##   [3]        -1         8        10</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_center</a></span>(rng), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -2         7        10
##   [2]        -3         6        10
##   [3]         1        10        10</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_3p</a></span>(grng), <span class="dv">10</span>) <span class="co"># leave negative strand fixed</span></a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [-4,  5]      +
##   [2]     seq1  [-7,  2]      *
##   [3]     seq1  [ 3, 12]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_5p</a></span>(grng), <span class="dv">10</span>) <span class="co"># leave positve strand fixed</span></a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [ 1, 10]      +
##   [2]     seq1  [ 2, 11]      *
##   [3]     seq1  [-1,  8]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Similarly, you can modify the width of an interval using the <code>stretch</code> verb. Without anchoring, this function will extend the interval in either direction by an integer amount. With anchoring, either the start, end or midpoint are preserved.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">rng2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_center</a></span>(rng), <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">rng2</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -7        13        21
##   [2]        -8        12        21
##   [3]        -5        15        21</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_end</a></span>(rng2), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]       -17        13        31
##   [2]       -18        12        31
##   [3]       -15        15        31</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_start</a></span>(rng2), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -7        23        31
##   [2]        -8        22        31
##   [3]        -5        25        31</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_3p</a></span>(grng), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
## Anchored by: 3p
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [-9, 15]      +
##   [2]     seq1  [-8, 12]      *
##   [3]     seq1  [-7, 18]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/ranges-anchor.html">anchor_5p</a></span>(grng), <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
## Anchored by: 5p
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [-9, 15]      +
##   [2]     seq1  [-8, 12]      *
##   [3]     seq1  [-7, 18]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>Ranges</code> can be shifted left or right. If strand information is available we can also shift upstream or downstream.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw"><a href="../reference/shift-ranges.html">shift_left</a></span>(rng, <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -9        -5         5
##   [2]        -8        -8         1
##   [3]        -7        -2         6</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw"><a href="../reference/shift-ranges.html">shift_right</a></span>(rng, <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        11        15         5
##   [2]        12        12         1
##   [3]        13        18         6</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="kw"><a href="../reference/shift-ranges.html">shift_upstream</a></span>(grng, <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [-9, -5]      +
##   [2]     seq1  [ 2,  2]      *
##   [3]     seq1  [13, 18]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="../reference/shift-ranges.html">shift_downstream</a></span>(grng, <span class="dv">10</span>)</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [11, 15]      +
##   [2]     seq1  [ 2,  2]      *
##   [3]     seq1  [-7, -2]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</section><section id="grouping-ranges" class="level1"><h1>Grouping <code>Ranges</code>
</h1>
<p><code>plyranges</code> introduces a new class of <code>Ranges</code> called <code>RangesGrouped</code>, this is a similiar idea to the grouped <code>data.frame\tibble</code> in <code>dplyr</code>.</p>
<p>Grouping can act on either the core components or the metadata columns of a <code>Ranges</code> object.</p>
<p>It is most effective when combined with other verbs such as <code><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code>, <code><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code>, <code><a href="http://dplyr.tidyverse.org/reference/filter.html">filter()</a></code>, <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code> or <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1">grng &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb35-2" data-line-number="2">         <span class="dt">strand =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),</a>
<a class="sourceLine" id="cb35-3" data-line-number="3">         <span class="dt">gc =</span> <span class="kw">runif</span>(<span class="dv">7</span>),</a>
<a class="sourceLine" id="cb35-4" data-line-number="4">         <span class="dt">start =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>,</a>
<a class="sourceLine" id="cb35-5" data-line-number="5">         <span class="dt">width =</span> <span class="dv">10</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb35-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</a>
<a class="sourceLine" id="cb35-7" data-line-number="7"></a>
<a class="sourceLine" id="cb35-8" data-line-number="8">grng_by_strand &lt;-<span class="st"> </span>grng <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb35-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(strand)</a>
<a class="sourceLine" id="cb35-10" data-line-number="10"></a>
<a class="sourceLine" id="cb35-11" data-line-number="11">grng_by_strand</a></code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
## Groups: strand
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2   [1, 10]      - | 0.889453538926318
##   [2]     chr2   [2, 11]      + | 0.180407245177776
##   [3]     chr2   [3, 12]      + | 0.629390850430354
##   [4]     chr1   [4, 13]      - | 0.989564136601985
##   [5]     chr1   [5, 14]      + | 0.130288870073855
##   [6]     chr2   [6, 15]      - | 0.330660525709391
##   [7]     chr2   [7, 16]      - | 0.865120546659455
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</section><section id="restricting-ranges" class="level1"><h1>Restricting <code>Ranges</code>
</h1>
<p>The verb <code>filter</code> can be used to restrict rows in the <code>Ranges</code>. Note that grouping will cause the <code>filter</code> to act within each group of the data.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">grng <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(gc <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.3</span>)</a></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2   [2, 11]      + | 0.180407245177776
##   [2]     chr1   [5, 14]      + | 0.130288870073855
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="co"># filtering by group </span></a>
<a class="sourceLine" id="cb39-2" data-line-number="2">grng_by_strand <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(gc <span class="op">==</span><span class="st"> </span><span class="kw">max</span>(gc)) </a></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
## Groups: strand
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2   [3, 12]      + | 0.629390850430354
##   [2]     chr1   [4, 13]      - | 0.989564136601985
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We also provide the convience methods <code>filter_by_overlaps</code> and <code>filter_by_non_overlaps</code> for restricting by any overlapping <code>Ranges</code>.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1">ir0 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>, <span class="dv">15</span>,<span class="dv">20</span>), <span class="dt">width =</span> <span class="dv">5</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb41-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>()</a>
<a class="sourceLine" id="cb41-3" data-line-number="3">ir1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">6</span>, <span class="dt">width =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">7</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb41-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_iranges</a></span>()</a>
<a class="sourceLine" id="cb41-5" data-line-number="5">ir0</a></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5
##   [3]        15        19         5
##   [4]        20        24         5</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1">ir1</a></code></pre></div>
<pre><code>## IRanges object with 5 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         4         3
##   [2]         3         6         4
##   [3]         4         8         5
##   [4]         5        10         6
##   [5]         6        12         7</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">ir0 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-filter-overlaps.html">filter_by_overlaps</a></span>(ir1)</a></code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1">ir0 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-filter-overlaps.html">filter_by_non_overlaps</a></span>(ir1) </a></code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        15        19         5
##   [2]        20        24         5</code></pre>
</section><section id="summarising-ranges" class="level1"><h1>Summarising <code>Ranges</code>
</h1>
<p>The <code>summarise</code> function will return a tibble because the information required to return a <code>Ranges</code> object is lost. It is often most useful to use <code>summarise</code> in combination with the <code>group_by</code> family of functions.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1">ir1 &lt;-<span class="st"> </span>ir1 <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb49-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">gc =</span> <span class="kw">runif</span>(<span class="kw">length</span>(.)))</a>
<a class="sourceLine" id="cb49-3" data-line-number="3"></a>
<a class="sourceLine" id="cb49-4" data-line-number="4">ir0 <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb49-5" data-line-number="5"><span class="st">  </span><span class="kw"><a href="../reference/ranges-overlaps.html">group_by_overlaps</a></span>(ir1) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb49-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">gc =</span> <span class="kw">mean</span>(gc))</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 2 columns
##       query        gc
##   &lt;integer&gt; &lt;numeric&gt;
## 1         1 0.6755545
## 2         2 0.6357952</code></pre>
</section><section id="joins-or-another-way-at-looking-at-overlaps-between-ranges" class="level1"><h1>Joins, or another way at looking at overlaps between <code>Ranges</code>
</h1>
<p>We can think about finding overlaps as a type of join between two <code>Ranges</code> where we are using the intervals as keys. Of most use are the <code>join_overlap_intersect</code>, <code>join_overlap_inner</code> and <code>join_overlap_left</code> functions.</p>
<p>To begin weâ€™ll define a query and subject <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">query &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"chr1"</span>,</a>
<a class="sourceLine" id="cb51-2" data-line-number="2">               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>),</a>
<a class="sourceLine" id="cb51-3" data-line-number="3">               <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">9</span>), </a>
<a class="sourceLine" id="cb51-4" data-line-number="4">               <span class="dt">end =</span>  <span class="kw">c</span>(<span class="dv">7</span>, <span class="dv">10</span>),</a>
<a class="sourceLine" id="cb51-5" data-line-number="5">               <span class="dt">key.a =</span> letters[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb51-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</a>
<a class="sourceLine" id="cb51-7" data-line-number="7"></a>
<a class="sourceLine" id="cb51-8" data-line-number="8">subject &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"chr1"</span>,</a>
<a class="sourceLine" id="cb51-9" data-line-number="9">               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"-"</span>, <span class="st">"+"</span>),</a>
<a class="sourceLine" id="cb51-10" data-line-number="10">               <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">6</span>), </a>
<a class="sourceLine" id="cb51-11" data-line-number="11">               <span class="dt">end =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb51-12" data-line-number="12">               <span class="dt">key.b =</span> LETTERS[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb51-13" data-line-number="13"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</a></code></pre></div>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-12-1.png" alt="Query and Subject Ranges" width="672"><p class="caption">
Query and Subject Ranges
</p>
</div>
<p>The first function, <code>join_overlap_intersect</code> will return a <code>Ranges`` object  where the start, end, and width coordinates correspond to the amount of any overlap between the left and right input</code>Ranges`. It also returns any metadata in the subject range if the subject overlaps the query. Note that by default the join functions ignore strandedness.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1">intersect_rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_intersect</a></span>(query, subject)</a>
<a class="sourceLine" id="cb52-2" data-line-number="2">intersect_rng</a></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1    [2, 4]      + |        a        A
##   [2]     chr1    [6, 7]      + |        a        B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1">intersect_df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(intersect_rng)[, <span class="op">-</span><span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">7</span>)]</a>
<a class="sourceLine" id="cb54-2" data-line-number="2">intersect_df<span class="op">$</span>key &lt;-<span class="st"> "Intersect Join"</span></a>
<a class="sourceLine" id="cb54-3" data-line-number="3">melted_ranges &lt;-<span class="st"> </span><span class="kw">rbind</span>(query_df, subject_df, intersect_df)</a>
<a class="sourceLine" id="cb54-4" data-line-number="4">melted_ranges<span class="op">$</span>key &lt;-<span class="st"> </span><span class="kw">factor</span>(melted_ranges<span class="op">$</span>key, </a>
<a class="sourceLine" id="cb54-5" data-line-number="5">                              <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">"Query"</span>, <span class="st">"Subject"</span>, <span class="st">"Intersect Join"</span>))</a>
<a class="sourceLine" id="cb54-6" data-line-number="6"><span class="kw">ggplot</span>(melted_ranges, <span class="kw">aes</span>(<span class="dt">xmin =</span> start, <span class="dt">xmax =</span> end, <span class="dt">ymin =</span> <span class="dv">1</span>, <span class="dt">ymax =</span> <span class="dv">3</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_rect</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb54-8" data-line-number="8"><span class="st">  </span><span class="kw">facet_grid</span>(key <span class="op">~</span><span class="st"> </span>.) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-9" data-line-number="9"><span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dt">by =</span> <span class="dv">1</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-10" data-line-number="10"><span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Position"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-11" data-line-number="11"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.y =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb54-12" data-line-number="12">        <span class="dt">axis.ticks.y =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb54-13" data-line-number="13">        <span class="dt">axis.title.y =</span> <span class="kw">element_blank</span>())</a></code></pre></div>
<p><img src="an-introduction_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>The <code>join_overlap_inner</code> function will return the <code>Ranges</code> in the query that overlap any <code>Ranges</code> in the subject. Like the <code>join_overlap_intersect</code> function metadata of the subject <code>Range</code> is returned if it overlaps the query.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1">inner_rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(query, subject)</a>
<a class="sourceLine" id="cb55-2" data-line-number="2">inner_rng</a></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1    [1, 7]      + |        a        A
##   [2]     chr1    [1, 7]      + |        a        B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-16-1.png" alt="Inner Join" width="672"><p class="caption">
Inner Join
</p>
</div>
<p>We also provide a convienence method called <code>find_overlaps</code> that computes the same result as <code>join_overlap_inner</code>.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1"><span class="kw"><a href="../reference/ranges-overlaps.html">find_overlaps</a></span>(query, subject)</a></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1    [1, 7]      + |        a        A
##   [2]     chr1    [1, 7]      + |        a        B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>The <code>join_overlap_left</code> method will perform an outer left join.</p>
<p>First any overlaps that are found will be returned similar to <code>join_overlap_inner</code>. Then any non-overlapping ranges will be returned, with missing values on the metadata columns.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1">left_rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_left</a></span>(query, subject)</a>
<a class="sourceLine" id="cb59-2" data-line-number="2">left_rng</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames    ranges strand |    key.a    key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   [1]     chr1   [1,  7]      + |        a        A
##   [2]     chr1   [1,  7]      + |        a        B
##   [3]     chr1   [9, 10]      - |        b     &lt;NA&gt;
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/unnamed-chunk-19-1.png" alt="Inner Join" width="672"><p class="caption">
Inner Join
</p>
</div>
<p>Compared with <code>filter_by_overlaps</code> above, the overlap left join expands the <code>Ranges</code> to give information about each interval on the query <code>Ranges</code> that overlap those on the subject <code>Ranges</code> as well as the intervals on the left that do not overlap any range on the right.</p>
<section id="example-dealing-with-multimapping" class="level2"><h2>Example: dealing with multimapping</h2>
<p>This example is taken from the Bioconductor support <a href="https://support.bioconductor.org/p/100046/">site</a>.</p>
<p>We have two <code>Ranges</code> objects. The first contains single nucleotide positions corresponding to an intensity measurement such as a ChiP-seq experiment, while the other contains coordinates for two genes of interest.</p>
<p>We want to identify which positions in the <code>intensties</code> <code>Ranges</code> overlap the genes, where each row corresponds to a position that overlaps a single gene.</p>
<p>First we create the two <code>Ranges</code> objects</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1">intensities &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"VI"</span>,</a>
<a class="sourceLine" id="cb61-2" data-line-number="2">                          <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">3320</span><span class="op">:</span><span class="dv">3321</span>,<span class="dv">3330</span><span class="op">:</span><span class="dv">3331</span>,<span class="dv">3341</span><span class="op">:</span><span class="dv">3342</span>),</a>
<a class="sourceLine" id="cb61-3" data-line-number="3">                          <span class="dt">width =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb61-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</a>
<a class="sourceLine" id="cb61-5" data-line-number="5"></a>
<a class="sourceLine" id="cb61-6" data-line-number="6">intensities </a></code></pre></div>
<pre><code>## GRanges object with 6 ranges and 0 metadata columns:
##       seqnames       ranges strand
##          &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]       VI [3320, 3320]      *
##   [2]       VI [3321, 3321]      *
##   [3]       VI [3330, 3330]      *
##   [4]       VI [3331, 3331]      *
##   [5]       VI [3341, 3341]      *
##   [6]       VI [3342, 3342]      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1">genes &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="st">"VI"</span>, </a>
<a class="sourceLine" id="cb63-2" data-line-number="2">                    <span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">3322</span>, <span class="dv">3030</span>),</a>
<a class="sourceLine" id="cb63-3" data-line-number="3">                    <span class="dt">end =</span> <span class="kw">c</span>(<span class="dv">3846</span>, <span class="dv">3338</span>),</a>
<a class="sourceLine" id="cb63-4" data-line-number="4">                    <span class="dt">gene_id=</span><span class="kw">c</span>(<span class="st">"YFL064C"</span>, <span class="st">"YFL065C"</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb63-5" data-line-number="5"><span class="st">  </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>()</a>
<a class="sourceLine" id="cb63-6" data-line-number="6">                    </a>
<a class="sourceLine" id="cb63-7" data-line-number="7">genes</a></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames       ranges strand |  gene_id
##          &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   [1]       VI [3322, 3846]      * |  YFL064C
##   [2]       VI [3030, 3338]      * |  YFL065C
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Now to find where the positions overlap each gene, we can perform an overlap join. This will automatically carry over the gene_id information as well as their coordinates (we can drop those by only selecting the gene_id).</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1">olap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(intensities, genes) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb65-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/select.html">select</a></span>(gene_id)</a>
<a class="sourceLine" id="cb65-3" data-line-number="3">olap</a></code></pre></div>
<pre><code>## GRanges object with 8 ranges and 1 metadata column:
##       seqnames       ranges strand |  gene_id
##          &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   [1]       VI [3320, 3320]      * |  YFL065C
##   [2]       VI [3321, 3321]      * |  YFL065C
##   [3]       VI [3330, 3330]      * |  YFL065C
##   [4]       VI [3330, 3330]      * |  YFL064C
##   [5]       VI [3331, 3331]      * |  YFL065C
##   [6]       VI [3331, 3331]      * |  YFL064C
##   [7]       VI [3341, 3341]      * |  YFL064C
##   [8]       VI [3342, 3342]      * |  YFL064C
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Several positions match to both genes. We can count them using <code>summarise</code> and grouping by the <code>start</code> position:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1">olap <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb67-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(start) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb67-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">n =</span> <span class="kw"><a href="../reference/n.html">n</a></span>())</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##       start         n
##   &lt;integer&gt; &lt;integer&gt;
## 1      3320         1
## 2      3321         1
## 3      3330         2
## 4      3331         2
## 5      3341         1
## 6      3342         1</code></pre>
</section><section id="finding-your-neighbours" class="level2"><h2>Finding your neighbours</h2>
<p>We also provide methods for finding nearest, preceding or following <code>Ranges</code>.</p>
<!--- ![Nearest neighbour joins](neighbours.png) --->
<p>These methods are similar to the overlap joins described above.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" data-line-number="1"><span class="kw"><a href="../reference/ranges-nearest.html">join_nearest</a></span>(ir0, ir1)</a></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |                gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt;
##   [1]         5         9         5 | 0.780358511023223
##   [2]        10        14         5 | 0.780358511023223
##   [3]        15        19         5 | 0.780358511023223
##   [4]        20        24         5 | 0.780358511023223</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb71-1" data-line-number="1"><span class="kw"><a href="../reference/ranges-follow.html">join_follow</a></span>(ir0, ir1)</a></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |                gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt;
##   [1]         5         9         5 | 0.777584439376369
##   [2]        10        14         5 | 0.603324356488883
##   [3]        15        19         5 | 0.780358511023223
##   [4]        20        24         5 | 0.780358511023223</code></pre>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb73-1" data-line-number="1"><span class="kw"><a href="../reference/ranges-precede.html">join_precede</a></span>(ir0, ir1) <span class="co"># nothing precedes returns empty `Ranges`</span></a></code></pre></div>
<pre><code>## IRanges object with 0 ranges and 1 metadata column:
##        start       end     width |        gc
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt;</code></pre>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb75-1" data-line-number="1"><span class="kw"><a href="../reference/ranges-precede.html">join_precede</a></span>(ir1, ir0)</a></code></pre></div>
<pre><code>## IRanges object with 5 ranges and 1 metadata column:
##           start       end     width |                gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt;
##   [1]         2         4         3 | 0.777584439376369
##   [2]         3         6         4 | 0.827303449623287
##   [3]         4         8         5 | 0.603324356488883
##   [4]         5        10         6 | 0.491231821943074
##   [5]         6        12         7 | 0.780358511023223</code></pre>
<p>For <code>GRanges</code> objects by default strand is not considered when performing overlap joins. To include strand when finding overlaps use append the prefix <code>directed</code>. To restrict overlapping <code>Ranges</code> to those within the query range use the prefix <code>within</code>.</p>
</section><section id="grouping-by-overlaps" class="level2"><h2>Grouping by overlaps</h2>
<p>Itâ€™s also possible to group by overlaps. Using this approach we can count the number of overlaps that are greater than 0.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb77-1" data-line-number="1">grp_by_olap &lt;-<span class="st"> </span>ir0 <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb77-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="../reference/ranges-overlaps.html">group_by_overlaps</a></span>(ir1)</a>
<a class="sourceLine" id="cb77-3" data-line-number="3">grp_by_olap</a></code></pre></div>
<pre><code>## IRanges object with 6 ranges and 2 metadata columns:
## Groups: query
##           start       end     width |                gc     query
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt; &lt;integer&gt;
##   [1]         5         9         5 | 0.827303449623287         1
##   [2]         5         9         5 | 0.603324356488883         1
##   [3]         5         9         5 | 0.491231821943074         1
##   [4]         5         9         5 | 0.780358511023223         1
##   [5]        10        14         5 | 0.491231821943074         2
##   [6]        10        14         5 | 0.780358511023223         2</code></pre>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb79-1" data-line-number="1">grp_by_olap <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb79-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">n_overlaps =</span> <span class="kw"><a href="../reference/n.html">n</a></span>())</a></code></pre></div>
<pre><code>## IRanges object with 6 ranges and 3 metadata columns:
## Groups: query
##           start       end     width |                gc     query
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |         &lt;numeric&gt; &lt;integer&gt;
##   [1]         5         9         5 | 0.827303449623287         1
##   [2]         5         9         5 | 0.603324356488883         1
##   [3]         5         9         5 | 0.491231821943074         1
##   [4]         5         9         5 | 0.780358511023223         1
##   [5]        10        14         5 | 0.491231821943074         2
##   [6]        10        14         5 | 0.780358511023223         2
##       n_overlaps
##        &lt;integer&gt;
##   [1]          4
##   [2]          4
##   [3]          4
##   [4]          4
##   [5]          2
##   [6]          2</code></pre>
<p>Of course we can also add overlap counts via the <code>count_overlaps</code> function.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" data-line-number="1">ir0 <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb81-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">n_overlaps =</span> <span class="kw"><a href="../reference/ranges-count-overlaps.Rd.html">count_overlaps</a></span>(., ir1))</a></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width | n_overlaps
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |  &lt;integer&gt;
##   [1]         5         9         5 |          4
##   [2]        10        14         5 |          2
##   [3]        15        19         5 |          0
##   [4]        20        24         5 |          0</code></pre>
</section></section><section id="data-importoutput" class="level1"><h1>Data Import/Output</h1>
<p>We provide convienence functions via <code>rtracklayer</code> and <code>GenomicAlignments</code> for reading/writing the following data formats to/from <code>Ranges</code> objects.</p>
<table class="table">
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>File Format</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/io-bam-read.html">read_bam()</a></code></td>
<td>BAM</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/io-bed-read.html">read_bed()</a></code>/<code><a href="../reference/io-bed-write.html">write_bed()</a></code>
</td>
<td>BED</td>
</tr>
<tr class="odd">
<td>
<code>read_bedgraph()</code>/ <code>write_bedgraph()</code>
</td>
<td>BEDGraph</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/io-bed-read.html">read_narrowpeaks()</a></code>/ <code><a href="../reference/io-bed-write.html">write_narrowpeaks()</a></code>
</td>
<td>narrowPeaks</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/io-gff-read.html">read_gff()</a></code> / <code><a href="../reference/io-gff-write.html">write_gff()</a></code>
</td>
<td>GFF(1-3) / GTF</td>
</tr>
<tr class="even">
<td>
<code>read_bw()</code> / <code>write_bw()</code>
</td>
<td>BigWig</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/io-wig-read.html">read_wig()</a></code> / <code><a href="../reference/io-wig-write.html">write_wig()</a></code>
</td>
<td>Wig</td>
</tr>
</tbody>
</table></section><section id="mapping-to-genomicrangesiranges" class="level1"><h1>Mapping to GenomicRanges/IRanges</h1>
<p>For users already familiar with the <code>IRanges</code> and <code>GenomicRanges</code> we provide mappings to the <code>plyranges</code> API.</p>
<section id="operations-on-range-width" class="level2"><h2>Operations on range width</h2>
<p>For <code>G</code>Ranges`` objects all functions ignore any strandedness, unless the strand of the range is anchored.</p>
<table class="table">
<colgroup>
<col style="width: 33%">
<col style="width: 18%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>anchor_(start/end/center/3p/5p)</code></td>
<td>Fix the <code>start/end/center/</code> coordinates or positive/negative strand of range. Can be used in combination with any of the following</td>
<td>Available in functions that have a <code>fix</code> argument.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/set_width.html">set_width(x, width)</a></code></td>
<td>Modify the width of a <code>Ranges</code>
</td>
<td><code>resize</code></td>
</tr>
<tr class="odd">
<td><code><a href="../reference/stretch.html">stretch(x, extend)</a></code></td>
<td>Extend the start and end coordinates in opposite directions by a fixed amount.</td>
<td>
<code>start(x)&lt;- start(x) + extend%/%2</code>, <code>end(x) &lt;- end(x) -extend%/%2</code>
</td>
</tr>
</tbody>
</table></section><section id="operations-on-range-width-invariant" class="level2"><h2>Operations on range width (invariant)</h2>
<table class="table">
<colgroup>
<col style="width: 33%">
<col style="width: 18%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>shift_[left/right/downstream/upstream](x, shift)</code></td>
<td>Shift the coordinates of the interval (left/right/downstream/upstream) by an integer amount.</td>
<td>
<code>shift_right</code> corresponds to <code>shift</code>
</td>
</tr>
<tr class="even">
<td><code>flank_[left/right/downstream/upstream](x, width)</code></td>
<td>Generates flanking regions of size width <code>left/right/downstream/upstream/</code>
</td>
<td>corresponds to <code>flank</code>
</td>
</tr>
</tbody>
</table></section><section id="set-operations-vector-wise" class="level2"><h2>Set operations (vector wise)</h2>
<p>These are usual set-operations that act on the sets of the <code>Ranges</code> represented in x and y. By default these operations will ignore any strand information. The directed versions of these functions will take into account strand.</p>
<table class="table">
<colgroup>
<col style="width: 33%">
<col style="width: 18%">
<col style="width: 47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>[intersect/setdiff/union/]_Ranges</code></td>
<td>Set operations between two <code>Ranges</code>, ignoring strand.</td>
<td>
<code>intersect/setdiff/union/</code> with <code>ignore.strand = FALSE</code>
</td>
</tr>
<tr class="even">
<td><code>[intersect/setdiff/union/]_anchored_Ranges</code></td>
<td>As above taking into account strandedness.</td>
<td></td>
</tr>
</tbody>
</table></section><section id="set-operations-element-wise" class="level2"><h2>Set operations (element wise)</h2>
<p>We provide infix operators and the verbs between and span to the represent element wise range operations. These map to the <code>pintersect/punion/psetdiff/pgap/punion(fill.gap = FALSE)</code> functions.</p>
</section><section id="restrictions" class="level2"><h2>Restrictions</h2>
<p>The verb <code>filter</code> corresponds to <code>subset</code>, while <code>filter_by_[overlaps/non_overlaps]</code> corresponds to <code>subsetByOverlaps</code>.</p>
</section><section id="aggregation" class="level2"><h2>Aggregation</h2>
<p>The <code>summarise</code> verb is most similar to the <code>aggregate</code> methods defined in <code>GenomicRanges/IRanges</code>.</p>
<p>The <code>reduce_ranges/disjoin_ranges</code> correspond to the <code>reduce/disjoin</code> methods. However, the former methods allow additional summarisation.</p>
<p>The <code><a href="../reference/compute_coverage.html">compute_coverage(x)</a></code> method corresponds to <code>[I/G]Ranges(coverage(x))</code>.</p>
</section><section id="overlaps" class="level2"><h2>Overlaps</h2>
<p>For <code>GRanges</code> objects all functions ignore any strandedness, unless the suffix <code>directed</code> is added to the funciton call</p>
<table class="table">
<colgroup>
<col style="width: 25%">
<col style="width: 55%">
<col style="width: 19%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> function</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/ranges-overlaps.html">find_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns a <code>Ranges</code> object with any range in <code>y</code> that overlaps <code>x</code>. Appends the metadata in <code>y</code> and its genomic intervals to the returning <code>Ranges</code>.</td>
<td>
<code>findOverlaps(x,y, maxgap, minoverlap, type = "any")</code> with expanding <code>x</code> and <code>y</code> by their hits and appending the <code>mcols</code> in <code>y</code>.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/ranges-overlaps.html">group_by_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns a GroupedRanges object grouped by the query hits.</td>
<td>Same as above with an additional column called <code>query</code> which contains the queryHits.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/ranges-count-overlaps.Rd.html">count_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns an integer vector (used with <code>mutate</code>)</td>
<td><code>countOverlaps(x, y, maxgap, minoverlap, type = "any")</code></td>
</tr>
<tr class="even">
<td><code><a href="../reference/ranges-overlaps-self.html">join_overlap_self(x, maxgap, minoverlap)</a></code></td>
<td>Returns a <code>Ranges</code> object with any range that overlaps itself.</td>
<td><code>findOverlaps(x,x, maxgap, minoverlap, type = "any")</code></td>
</tr>
<tr class="odd">
<td><code><a href="../reference/overlap-joins.html">join_overlap_inner(x, y, maxgap, minoverlap)</a></code></td>
<td>Finds the intersecting <code>Ranges</code> that overlap in <code>x</code> and <code>y</code>. Returns a <code>Ranges</code> object with the metadata from <code>x</code> and <code>y</code>.</td>
<td>
<code>findOverlapsPairs(x,y, maxgap, minoverlap, type = "any")</code> + <code>pintersect</code>.</td>
</tr>
<tr class="even">
<td><code>*_within</code></td>
<td>Adding suffix <code>within</code> will find overlaps</td>
<td>Makes <code>type = "within"</code>
</td>
</tr>
<tr class="odd">
<td><code>*_includes</code></td>
<td>inverse of within functions</td>
<td>-</td>
</tr>
<tr class="even">
<td><code>join_nearest[_left/right/up/downstream](x,y)</code></td>
<td>Finds nearest neighbour <code>Ranges</code> between <code>x</code> and <code>y</code>.</td>
<td>
<code>nearest</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
<tr class="odd">
<td><code>join_precede[_left/right/up/downstream](x,y)</code></td>
<td>Finds <code>Ranges</code> in <code>x</code> that preced <code>y</code>
</td>
<td>
<code>precedes</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
<tr class="even">
<td><code>join_follow[_left/right/up/downstream](x,y)</code></td>
<td>Finds <code>Ranges</code> in <code>x</code> that follow <code>y</code>
</td>
<td>
<code>precedes</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
</tbody>
</table></section></section><section id="appendix" class="level1"><h1>Appendix</h1>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 3.4.3 (2017-11-30)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] ggplot2_2.2.1        plyranges_0.99.4     GenomicRanges_1.30.3
## [4] GenomeInfoDb_1.14.0  IRanges_2.12.0       S4Vectors_0.16.0    
## [7] BiocGenerics_0.24.0 
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.15               highr_0.6                 
##  [3] plyr_1.8.4                 pillar_1.2.1              
##  [5] compiler_3.4.3             XVector_0.18.0            
##  [7] bindr_0.1                  bitops_1.0-6              
##  [9] tools_3.4.3                zlibbioc_1.24.0           
## [11] digest_0.6.15              gtable_0.2.0              
## [13] tibble_1.4.2               evaluate_0.10.1           
## [15] lattice_0.20-35            pkgconfig_2.0.1           
## [17] rlang_0.2.0.9000           Matrix_1.2-12             
## [19] DelayedArray_0.4.1         yaml_2.1.16               
## [21] bindrcpp_0.2               GenomeInfoDbData_1.0.0    
## [23] rtracklayer_1.38.3         stringr_1.3.0             
## [25] dplyr_0.7.4                knitr_1.18                
## [27] Biostrings_2.46.0          tidyselect_0.2.4          
## [29] rprojroot_1.3-2            grid_3.4.3                
## [31] glue_1.2.0                 Biobase_2.38.0            
## [33] R6_2.2.2                   XML_3.98-1.9              
## [35] BiocParallel_1.12.0        rmarkdown_1.8             
## [37] reshape2_1.4.3             purrr_0.2.4               
## [39] tidyr_0.8.0                magrittr_1.5              
## [41] scales_0.5.0               backports_1.1.2           
## [43] Rsamtools_1.30.0           htmltools_0.3.6           
## [45] matrixStats_0.53.0         GenomicAlignments_1.14.1  
## [47] assertthat_0.2.0           SummarizedExperiment_1.8.1
## [49] colorspace_1.3-2           labeling_0.3              
## [51] stringi_1.1.6              lazyeval_0.2.1            
## [53] munsell_0.4.3              RCurl_1.95-4.10</code></pre>
</section>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#ranges-revisted"><code>Ranges</code> revisted</a></li>
      <li><a href="#constructing-ranges">Constructing <code>Ranges</code></a></li>
      <li><a href="#arithmetic-on-ranges">Arithmetic on Ranges</a></li>
      <li><a href="#grouping-ranges">Grouping <code>Ranges</code></a></li>
      <li><a href="#restricting-ranges">Restricting <code>Ranges</code></a></li>
      <li><a href="#summarising-ranges">Summarising <code>Ranges</code></a></li>
      <li>
<a href="#joins-or-another-way-at-looking-at-overlaps-between-ranges">Joins, or another way at looking at overlaps between <code>Ranges</code></a><ul class="nav nav-pills nav-stacked">
<li><a href="#example-dealing-with-multimapping">Example: dealing with multimapping</a></li>
      <li><a href="#finding-your-neighbours">Finding your neighbours</a></li>
      <li><a href="#grouping-by-overlaps">Grouping by overlaps</a></li>
      </ul>
</li>
      <li><a href="#data-importoutput">Data Import/Output</a></li>
      <li>
<a href="#mapping-to-genomicrangesiranges">Mapping to GenomicRanges/IRanges</a><ul class="nav nav-pills nav-stacked">
<li><a href="#operations-on-range-width">Operations on range width</a></li>
      <li><a href="#operations-on-range-width-invariant">Operations on range width (invariant)</a></li>
      <li><a href="#set-operations-vector-wise">Set operations (vector wise)</a></li>
      <li><a href="#set-operations-element-wise">Set operations (element wise)</a></li>
      <li><a href="#restrictions">Restrictions</a></li>
      <li><a href="#aggregation">Aggregation</a></li>
      <li><a href="#overlaps">Overlaps</a></li>
      </ul>
</li>
      <li><a href="#appendix">Appendix</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://stuartlee.org">Stuart Lee</a>, <a href="https://github.com/lawremi">Michael Lawrence</a>, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
