<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using plyranges and GenomicRanges to perform common genomic data wrangling tasks • plyranges</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using plyranges and GenomicRanges to perform common genomic data wrangling tasks">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plyranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/an-introduction.html">Introduction to plyranges</a>
    </li>
    <li>
      <a href="../articles/common-tasks.html">Common genomic data wrangling tasks</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/plyranges">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using plyranges and GenomicRanges to perform common genomic data wrangling tasks</h1>
                        <h4 class="author">Stuart Lee and Michael Lawrence</h4>
            
            <h4 class="date">2018-11-27</h4>
      
      
      <div class="hidden name"><code>common-tasks.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This tutorial will show you how to think about and perform common genomic data wrangling tasks using the using the <a href="">plyranges</a> and <a href="">GenomicRanges</a> Bioconductor packages. These packages represent genomic data using the <em>GRanges</em> data structure, which provides a powerful abstraction for working with measurements or annotations along genomic regions. In this tutorial you will use <em>GRanges</em> object to construct workflows to</p>
<ul>
<li>read data from common formats like BED</li>
<li>find overlaps between different sets of genomic regions</li>
<li>compute coverage and related summaries</li>
</ul>
<div id="who-is-this-tutorial-is-for" class="section level2">
<h2 class="hasAnchor">
<a href="#who-is-this-tutorial-is-for" class="anchor"></a>Who is this tutorial is for?</h2>
<p>If you are an R beginner or new to the Bioconductor ecosystem, this tutorial provides a comprehensive overview of one of the most important data structures in Bioconductor, <em>GRanges</em>. We show you how to think of this data structure as a special kind of table and perform all kinds of operations on it.</p>
<p>If you are familiar with the tidyverse suite of packages, then the operations and workflows we will develop will look somewhat familiar to you. However, due to the complexity of genomics data, plyranges develops a new grammar based on the tidyverse to meet the needs of the computational biologist.</p>
<p>We assume you know what a data.frame is and have some knowledge of vector types in R.</p>
<p>The examples and and exercises are based on the <a href="http://bioconductor.org/packages/release/bioc/vignettes/HelloRanges/inst/doc/tutorial.pdf">HelloRanges package tutorial</a>. If you are familiar with bedtools, this is an excellent resource for learning how similar operations are performed with Bioconductor packages.</p>
</div>
<div id="setup" class="section level2">
<h2 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h2>
<p>You will need R (version 3.5 or greater) with the following packages installed:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Bioconductor</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(GenomicRanges)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(plyranges)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(HelloRangesData)</code></pre></div>
<p>If you do not have these packages you can install them as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="co"># -&gt; Bioconductor packages are installed with this package</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="co"># then use the install function from BiocManager </span>
BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"GenomicRanges"</span>, <span class="st">"plyranges"</span>, <span class="st">"HelloRangesData"</span>))</code></pre></div>
</div>
</div>
<div id="introducing-granges" class="section level1">
<h1 class="hasAnchor">
<a href="#introducing-granges" class="anchor"></a>Introducing GRanges</h1>
<p>The GRanges is a fundamental data structure used in the Bioconductor ecosystem. It is very similar to the base R data.frame but with appropriate semantics for a genomics experiment: it has fixed columns for the chromosome, start and end coordinates, and the strand, along with an arbitrary set of additional columns, consisting of measurements or metadata specific to the data type or experiment.</p>
<div class="figure">
<img src="GRanges.png" alt="An illustration of a GRanges data object for a single sample from an RNA-seq experiment. The core components of the object include a seqnames column (representing the chromosome), a ranges column which consists of start and end coordinates for a genomic region, and a strand identifier (either positive, negative, or unstranded). Metadata are included as columns to the right of the dotted line as annotations (gene-id) or range level covariates (score)." width="\textwidth"><p class="caption">
An illustration of a GRanges data object for a single sample from an RNA-seq experiment. The core components of the object include a seqnames column (representing the chromosome), a ranges column which consists of start and end coordinates for a genomic region, and a strand identifier (either positive, negative, or unstranded). Metadata are included as columns to the right of the dotted line as annotations (gene-id) or range level covariates (score).
</p>
</div>
<p>GRanges balances flexibility with formal constraints, so that it is applicable to virtually any genomic workflow, while also being semantically rich enough to support high-level operations on genomic ranges.</p>
<p>We can create a GRanges from an ordinary data.frame with plyranges using the <code><a href="../reference/ranges-construct.html">as_granges()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(plyranges, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">seqnames =</span> <span class="st">"VI"</span>, 
                    <span class="dt">start =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">3322</span>, <span class="dv">3030</span>, <span class="dv">1437</span>,  <span class="dv">5066</span>, <span class="dv">6426</span>, <span class="dv">836</span>),
                    <span class="dt">end =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">3846</span>, <span class="dv">3338</span>, <span class="dv">2615</span>, <span class="dv">5521</span>, <span class="dv">7565</span>, <span class="dv">1363</span>),
                    <span class="dt">strand =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>),
                    <span class="dt">gene_id=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"YFL064C"</span>, <span class="st">"YFL065C"</span>, <span class="st">"YFL066C"</span>, 
                              <span class="st">"YFL063W"</span>, <span class="st">"YFL062W"</span>, <span class="st">"YFL067W"</span>),
                    <span class="dt">gc =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dv">6</span>),
                    <span class="dt">score =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(<span class="dv">6</span>, <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">p =</span> <span class="fl">0.1</span>),
                    <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-construct.html">as_granges</a></span>(genes)

genes</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 3 metadata columns:
##       seqnames    ranges strand |     gene_id                gc     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;numeric&gt; &lt;integer&gt;
##   [1]       VI 3322-3846      - |     YFL064C 0.406736710341647         0
##   [2]       VI 3030-3338      - |     YFL065C 0.449663328705356         1
##   [3]       VI 1437-2615      - |     YFL066C 0.868391741998494         0
##   [4]       VI 5066-5521      + |     YFL063W 0.935484018875286         0
##   [5]       VI 6426-7565      + |     YFL062W 0.896344820735976         0
##   [6]       VI  836-1363      + |     YFL067W 0.765752661740407         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>You might notice a few differences between how the <code>genes</code> GRanges object is printed compared to an ordinary data.frame or tibble.</p>
<p>The measurements/annotation columns like gene_id, gc and score are placed on the right hand side of the “|”. These are referred to as metadata columns; they are not strictly required to construct a GRanges object but are useful for providing biological context.</p>
<p>The required part of a GRanges object are left of the “|” character. These are the seqnames column, the ranges (start, end, width) column, and the strand column.</p>
<p>You may have also noticed that underneath each column name is the type of that column. The metadata columnns are of some usual base R types like <code>character</code>, <code>numeric</code>, and <code>integer</code> but the core columns may be unfamilar. The seqnames and strand column are (factor) <code>Rle</code> columns. <code>Rle</code> stands for run length encoding, and refers to how a column is stored. For now, there is no harm in thinking about these as ordinary factors.</p>
<p>The ranges column is an <code>IRanges</code> which really consists of three columns: the start, end and width coordinates of the range. This becomes apparent if you look back to our original data.frame - the start and end columns from there are just the start and end intervals of the ranges column.</p>
<p>Do not be concerned if this a bit confusing for now, once we start playing with our GRanges object, everything will become clearer.</p>
</div>
<div id="granges-transformation-and-import" class="section level1">
<h1 class="hasAnchor">
<a href="#granges-transformation-and-import" class="anchor"></a>GRanges transformation and import</h1>
<p>Once you have obtained data from an experiement or a public repository, you often want to compute summaries, create new columns, modify the genomic regions or just rearrange the data. Here you will learn how to achieve these tasks using the plyranges:</p>
<ul>
<li>Modify, subset, and aggregate genomic data with the <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>, <code><a href="../reference/tidyverse-reexports.html">filter()</a></code>, and <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code>.</li>
<li>Modify genomic regions with the <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>…</li>
<li>And while fixing the start/end/center coordinates with the <code>anchor_*</code> family of functions.</li>
<li>Merge all overlapping and adjacent genomic regions with <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code>.</li>
<li>Merge the end points of all genomic regions with <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</li>
<li>Sort genomic ranges with <code><a href="../reference/tidyverse-reexports.html">arrange()</a></code>.</li>
</ul>
<p>Any of the above operations can be performed on subsets of the data with <code><a href="../reference/tidyverse-reexports.html">group_by()</a></code>.</p>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<p>To begin we will test out some of these operations using a GRanges object obtained from a BED file containing annotated exons from RefSeq from the hg19 genome build. To load the file as GRanges we can simply use the <code>read_bed</code> function from plyranges.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons_bed &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"exons.bed"</span>, <span class="dt">package=</span><span class="st">"HelloRangesData"</span>)

<span class="co"># our genome build, requires internet connection</span>
build &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-info.html">genome_info</a></span>(<span class="st">"hg19"</span>)

build</code></pre></div>
<pre><code>## GRanges object with 93 ranges and 0 metadata columns:
##                        seqnames      ranges strand
##                           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##             chr1           chr1 1-249250621      *
##             chr2           chr2 1-243199373      *
##             chr3           chr3 1-198022430      *
##             chr4           chr4 1-191154276      *
##             chr5           chr5 1-180915260      *
##              ...            ...         ...    ...
##   chrUn_gl000245 chrUn_gl000245     1-36651      *
##   chrUn_gl000246 chrUn_gl000246     1-38154      *
##   chrUn_gl000247 chrUn_gl000247     1-36422      *
##   chrUn_gl000248 chrUn_gl000248     1-39786      *
##   chrUn_gl000249 chrUn_gl000249     1-38502      *
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons &lt;-<span class="st"> </span><span class="kw"><a href="../reference/io-bed-read.html">read_bed</a></span>(exons_bed, 
                  <span class="dt">genome_info =</span> build)
exons</code></pre></div>
<pre><code>## GRanges object with 459752 ranges and 2 metadata columns:
##            seqnames            ranges strand |
##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |
##        [1]     chr1       11874-12227      + |
##        [2]     chr1       12613-12721      + |
##        [3]     chr1       13221-14409      + |
##        [4]     chr1       14362-14829      - |
##        [5]     chr1       14970-15038      - |
##        ...      ...               ...    ... .
##   [459748]     chrY 59338754-59338859      + |
##   [459749]     chrY 59338754-59338859      + |
##   [459750]     chrY 59340194-59340278      + |
##   [459751]     chrY 59342487-59343488      + |
##   [459752]     chrY 59342487-59343488      + |
##                                          name     score
##                                   &lt;character&gt; &lt;numeric&gt;
##        [1]    NR_046018_exon_0_0_chr1_11874_f         0
##        [2]    NR_046018_exon_1_0_chr1_12613_f         0
##        [3]    NR_046018_exon_2_0_chr1_13221_f         0
##        [4]    NR_024540_exon_0_0_chr1_14362_r         0
##        [5]    NR_024540_exon_1_0_chr1_14970_r         0
##        ...                                ...       ...
##   [459748] NM_002186_exon_6_0_chrY_59338754_f         0
##   [459749] NM_176786_exon_7_0_chrY_59338754_f         0
##   [459750] NM_002186_exon_7_0_chrY_59340194_f         0
##   [459751] NM_002186_exon_8_0_chrY_59342487_f         0
##   [459752] NM_176786_exon_8_0_chrY_59342487_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>The <code><a href="../reference/ranges-info.html">genome_info()</a></code> function automatically propagates the reference annotations from the hg19 reference as a GRanges. This is important for ensuring the integrity of any operations involving modifications to genomic regions, like overlaps or nearest neighbours.</p>
</div>
<div id="filter-ranges-with-filter" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-ranges-with-filter" class="anchor"></a>Filter ranges with <code><a href="../reference/tidyverse-reexports.html">filter()</a></code>
</h2>
<p>We can <code><a href="../reference/tidyverse-reexports.html">filter()</a></code> ranges based on column values being <code>TRUE</code>. This requries passing the ranges, along with one or more logical expressions.</p>
<p>For example, we can restrict exons to those that are on the the X chromosome:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(exons, seqnames <span class="op">==</span><span class="st"> "chrX"</span>)</code></pre></div>
<pre><code>## GRanges object with 18173 ranges and 2 metadata columns:
##           seqnames              ranges strand |
##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##       [1]     chrX       192991-193061      + |
##       [2]     chrX       198061-198351      + |
##       [3]     chrX       198149-198351      + |
##       [4]     chrX       200834-200981      + |
##       [5]     chrX       200834-200981      + |
##       ...      ...                 ...    ... .
##   [18169]     chrX 155235748-155235853      + |
##   [18170]     chrX 155235748-155235853      + |
##   [18171]     chrX 155237188-155237272      + |
##   [18172]     chrX 155239481-155240482      + |
##   [18173]     chrX 155239481-155240482      + |
##                                          name     score
##                                   &lt;character&gt; &lt;numeric&gt;
##       [1]    NR_028057_exon_0_0_chrX_192991_f         0
##       [2]    NM_018390_exon_0_0_chrX_198061_f         0
##       [3]    NR_028057_exon_1_0_chrX_198149_f         0
##       [4]    NM_018390_exon_1_0_chrX_200834_f         0
##       [5]    NR_028057_exon_2_0_chrX_200834_f         0
##       ...                                 ...       ...
##   [18169] NM_002186_exon_6_0_chrX_155235748_f         0
##   [18170] NM_176786_exon_7_0_chrX_155235748_f         0
##   [18171] NM_002186_exon_7_0_chrX_155237188_f         0
##   [18172] NM_002186_exon_8_0_chrX_155239481_f         0
##   [18173] NM_176786_exon_8_0_chrX_155239481_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Multiple conditions can be composed together within a call to filter by seperating the expressions with a comma. To restrict exons to those that are on the X chromosome <em>AND</em> on the positive strand:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(exons, seqnames <span class="op">==</span><span class="st"> "chrX"</span>, strand <span class="op">==</span><span class="st"> "+"</span>)</code></pre></div>
<pre><code>## GRanges object with 9176 ranges and 2 metadata columns:
##          seqnames              ranges strand |
##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##      [1]     chrX       192991-193061      + |
##      [2]     chrX       198061-198351      + |
##      [3]     chrX       198149-198351      + |
##      [4]     chrX       200834-200981      + |
##      [5]     chrX       200834-200981      + |
##      ...      ...                 ...    ... .
##   [9172]     chrX 155235748-155235853      + |
##   [9173]     chrX 155235748-155235853      + |
##   [9174]     chrX 155237188-155237272      + |
##   [9175]     chrX 155239481-155240482      + |
##   [9176]     chrX 155239481-155240482      + |
##                                         name     score
##                                  &lt;character&gt; &lt;numeric&gt;
##      [1]    NR_028057_exon_0_0_chrX_192991_f         0
##      [2]    NM_018390_exon_0_0_chrX_198061_f         0
##      [3]    NR_028057_exon_1_0_chrX_198149_f         0
##      [4]    NM_018390_exon_1_0_chrX_200834_f         0
##      [5]    NR_028057_exon_2_0_chrX_200834_f         0
##      ...                                 ...       ...
##   [9172] NM_002186_exon_6_0_chrX_155235748_f         0
##   [9173] NM_176786_exon_7_0_chrX_155235748_f         0
##   [9174] NM_002186_exon_7_0_chrX_155237188_f         0
##   [9175] NM_002186_exon_8_0_chrX_155239481_f         0
##   [9176] NM_176786_exon_8_0_chrX_155239481_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>This is equivalent to</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(exons, seqnames <span class="op">==</span><span class="st"> "chrX"</span> <span class="op">&amp;</span><span class="st"> </span>strand <span class="op">==</span><span class="st"> "+"</span>)</code></pre></div>
<pre><code>## GRanges object with 9176 ranges and 2 metadata columns:
##          seqnames              ranges strand |
##             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##      [1]     chrX       192991-193061      + |
##      [2]     chrX       198061-198351      + |
##      [3]     chrX       198149-198351      + |
##      [4]     chrX       200834-200981      + |
##      [5]     chrX       200834-200981      + |
##      ...      ...                 ...    ... .
##   [9172]     chrX 155235748-155235853      + |
##   [9173]     chrX 155235748-155235853      + |
##   [9174]     chrX 155237188-155237272      + |
##   [9175]     chrX 155239481-155240482      + |
##   [9176]     chrX 155239481-155240482      + |
##                                         name     score
##                                  &lt;character&gt; &lt;numeric&gt;
##      [1]    NR_028057_exon_0_0_chrX_192991_f         0
##      [2]    NM_018390_exon_0_0_chrX_198061_f         0
##      [3]    NR_028057_exon_1_0_chrX_198149_f         0
##      [4]    NM_018390_exon_1_0_chrX_200834_f         0
##      [5]    NR_028057_exon_2_0_chrX_200834_f         0
##      ...                                 ...       ...
##   [9172] NM_002186_exon_6_0_chrX_155235748_f         0
##   [9173] NM_176786_exon_7_0_chrX_155235748_f         0
##   [9174] NM_002186_exon_7_0_chrX_155237188_f         0
##   [9175] NM_002186_exon_8_0_chrX_155239481_f         0
##   [9176] NM_176786_exon_8_0_chrX_155239481_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>But not the same as filtering exons to those on the X chromosome or positively stranded.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(exons, seqnames <span class="op">==</span><span class="st"> "chrX"</span> <span class="op">|</span><span class="st"> </span>strand <span class="op">==</span><span class="st"> "+"</span>)</code></pre></div>
<pre><code>## GRanges object with 241162 ranges and 2 metadata columns:
##            seqnames            ranges strand |
##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |
##        [1]     chr1       11874-12227      + |
##        [2]     chr1       12613-12721      + |
##        [3]     chr1       13221-14409      + |
##        [4]     chr1       69091-70008      + |
##        [5]     chr1     323892-324060      + |
##        ...      ...               ...    ... .
##   [241158]     chrY 59338754-59338859      + |
##   [241159]     chrY 59338754-59338859      + |
##   [241160]     chrY 59340194-59340278      + |
##   [241161]     chrY 59342487-59343488      + |
##   [241162]     chrY 59342487-59343488      + |
##                                          name     score
##                                   &lt;character&gt; &lt;numeric&gt;
##        [1]    NR_046018_exon_0_0_chr1_11874_f         0
##        [2]    NR_046018_exon_1_0_chr1_12613_f         0
##        [3]    NR_046018_exon_2_0_chr1_13221_f         0
##        [4] NM_001005484_exon_0_0_chr1_69091_f         0
##        [5]   NR_028322_exon_0_0_chr1_323892_f         0
##        ...                                ...       ...
##   [241158] NM_002186_exon_6_0_chrY_59338754_f         0
##   [241159] NM_176786_exon_7_0_chrY_59338754_f         0
##   [241160] NM_002186_exon_7_0_chrY_59340194_f         0
##   [241161] NM_002186_exon_8_0_chrY_59342487_f         0
##   [241162] NM_176786_exon_8_0_chrY_59342487_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>For a comprehensive overview of all R’s logical operators see <a href="https://r4ds.had.co.nz/transform.html#filter-rows-with-filter"><strong>section 5.2.2 of R for Data Science</strong></a>.</p>
<div id="overlap-based-filtering" class="section level3">
<h3 class="hasAnchor">
<a href="#overlap-based-filtering" class="anchor"></a>Overlap based filtering</h3>
<p>We can also restrict ranges by overlaps using <code><a href="../reference/ranges-filter-overlaps.html">filter_by_overlaps()</a></code>, or exclude non overlapping features using <code><a href="../reference/ranges-filter-overlaps.html">filter_by_non_overlaps()</a></code>.</p>
<p>To see how these work it’ll be useful to load another GRanges object, this time from a BED file containing CpG islands.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg_bed &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"cpg.bed"</span>, <span class="dt">package =</span> <span class="st">"HelloRangesData"</span>)

cpg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/io-bed-read.html">read_bed</a></span>(cpg_bed, 
                <span class="dt">genome_info =</span> build)
cpg</code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 1 metadata column:
##           seqnames            ranges strand |        name
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116
##       [2]     chr1     135125-135563      * |     CpG:_30
##       [3]     chr1     327791-328229      * |     CpG:_29
##       [4]     chr1     437152-438164      * |     CpG:_84
##       [5]     chr1     449274-450544      * |     CpG:_99
##       ...      ...               ...    ... .         ...
##   [28687]     chrY 27610116-27611088      * |     CpG:_76
##   [28688]     chrY 28555536-28555932      * |     CpG:_32
##   [28689]     chrY 28773316-28773544      * |     CpG:_25
##   [28690]     chrY 59213795-59214183      * |     CpG:_36
##   [28691]     chrY 59349267-59349574      * |     CpG:_29
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>To find exons that do not overlap any cpg islands we</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-filter-overlaps.html">filter_by_non_overlaps</a></span>(exons, cpg)</code></pre></div>
<pre><code>## GRanges object with 414647 ranges and 2 metadata columns:
##            seqnames            ranges strand |
##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |
##        [1]     chr1       11874-12227      + |
##        [2]     chr1       12613-12721      + |
##        [3]     chr1       13221-14409      + |
##        [4]     chr1       14362-14829      - |
##        [5]     chr1       14970-15038      - |
##        ...      ...               ...    ... .
##   [414643]     chrY 59338754-59338859      + |
##   [414644]     chrY 59338754-59338859      + |
##   [414645]     chrY 59340194-59340278      + |
##   [414646]     chrY 59342487-59343488      + |
##   [414647]     chrY 59342487-59343488      + |
##                                          name     score
##                                   &lt;character&gt; &lt;numeric&gt;
##        [1]    NR_046018_exon_0_0_chr1_11874_f         0
##        [2]    NR_046018_exon_1_0_chr1_12613_f         0
##        [3]    NR_046018_exon_2_0_chr1_13221_f         0
##        [4]    NR_024540_exon_0_0_chr1_14362_r         0
##        [5]    NR_024540_exon_1_0_chr1_14970_r         0
##        ...                                ...       ...
##   [414643] NM_002186_exon_6_0_chrY_59338754_f         0
##   [414644] NM_176786_exon_7_0_chrY_59338754_f         0
##   [414645] NM_002186_exon_7_0_chrY_59340194_f         0
##   [414646] NM_002186_exon_8_0_chrY_59342487_f         0
##   [414647] NM_176786_exon_8_0_chrY_59342487_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Or we could find the cpg islands that are overlapped by exons</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ranges-filter-overlaps.html">filter_by_overlaps</a></span>(cpg, exons)</code></pre></div>
<pre><code>## GRanges object with 18845 ranges and 1 metadata column:
##           seqnames            ranges strand |        name
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116
##       [2]     chr1     135125-135563      * |     CpG:_30
##       [3]     chr1     327791-328229      * |     CpG:_29
##       [4]     chr1     713985-714547      * |     CpG:_60
##       [5]     chr1     762417-763445      * |    CpG:_115
##       ...      ...               ...    ... .         ...
##   [18841]     chrY 25345105-25345331      * |     CpG:_21
##   [18842]     chrY 25365504-25365730      * |     CpG:_21
##   [18843]     chrY 26959490-26959716      * |     CpG:_21
##   [18844]     chrY 26979890-26980116      * |     CpG:_21
##   [18845]     chrY 59213795-59214183      * |     CpG:_36
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
</div>
<div id="modify-metadata-with-mutate" class="section level2">
<h2 class="hasAnchor">
<a href="#modify-metadata-with-mutate" class="anchor"></a>Modify metadata with <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>
</h2>
<p>New metadata columns can be created using the <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code> function. The first argument is the GRanges object we are mutating and the remaining arguments are name-value pairs, where name refers to name of the new column being created and value refers to the value the new column will take.</p>
<p>As an example, we make a new column called total_length which contains the number of bases of in each CpG island. This is just a copy of the width column in Ranges object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(cpg, <span class="dt">total_length =</span>  width)</code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 2 metadata columns:
##           seqnames            ranges strand |        name total_length
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;integer&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116         1075
##       [2]     chr1     135125-135563      * |     CpG:_30          439
##       [3]     chr1     327791-328229      * |     CpG:_29          439
##       [4]     chr1     437152-438164      * |     CpG:_84         1013
##       [5]     chr1     449274-450544      * |     CpG:_99         1271
##       ...      ...               ...    ... .         ...          ...
##   [28687]     chrY 27610116-27611088      * |     CpG:_76          973
##   [28688]     chrY 28555536-28555932      * |     CpG:_32          397
##   [28689]     chrY 28773316-28773544      * |     CpG:_25          229
##   [28690]     chrY 59213795-59214183      * |     CpG:_36          389
##   [28691]     chrY 59349267-59349574      * |     CpG:_29          308
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Multiple columns can be computed inside of a call to <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code> by seperating each pair of name-value expressions with a comma, we could add an additional column called mid, which is the midpoint of the CpG island.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(cpg,
       <span class="dt">total_length =</span> width,
       <span class="dt">mid =</span> start <span class="op">+</span><span class="st"> </span>(width <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%/%</span><span class="st"> </span><span class="dv">2</span>
       )</code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 3 metadata columns:
##           seqnames            ranges strand |        name total_length
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;integer&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116         1075
##       [2]     chr1     135125-135563      * |     CpG:_30          439
##       [3]     chr1     327791-328229      * |     CpG:_29          439
##       [4]     chr1     437152-438164      * |     CpG:_84         1013
##       [5]     chr1     449274-450544      * |     CpG:_99         1271
##       ...      ...               ...    ... .         ...          ...
##   [28687]     chrY 27610116-27611088      * |     CpG:_76          973
##   [28688]     chrY 28555536-28555932      * |     CpG:_32          397
##   [28689]     chrY 28773316-28773544      * |     CpG:_25          229
##   [28690]     chrY 59213795-59214183      * |     CpG:_36          389
##   [28691]     chrY 59349267-59349574      * |     CpG:_29          308
##                 mid
##           &lt;numeric&gt;
##       [1]     29273
##       [2]    135344
##       [3]    328010
##       [4]    437658
##       [5]    449909
##       ...       ...
##   [28687]  27610602
##   [28688]  28555734
##   [28689]  28773430
##   [28690]  59213989
##   [28691]  59349420
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Moreover, <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code> keeps track of which columnns have been created, so you can create new columns using pre-existing ones.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(cpg,
       <span class="dt">total_length =</span> width,
       <span class="dt">mid =</span> start <span class="op">+</span><span class="st"> </span>(width <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%/%</span><span class="st"> </span><span class="dv">2</span>,
       <span class="dt">mid2 =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">sqrt</a></span>(mid)
       )</code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 4 metadata columns:
##           seqnames            ranges strand |        name total_length
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;integer&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116         1075
##       [2]     chr1     135125-135563      * |     CpG:_30          439
##       [3]     chr1     327791-328229      * |     CpG:_29          439
##       [4]     chr1     437152-438164      * |     CpG:_84         1013
##       [5]     chr1     449274-450544      * |     CpG:_99         1271
##       ...      ...               ...    ... .         ...          ...
##   [28687]     chrY 27610116-27611088      * |     CpG:_76          973
##   [28688]     chrY 28555536-28555932      * |     CpG:_32          397
##   [28689]     chrY 28773316-28773544      * |     CpG:_25          229
##   [28690]     chrY 59213795-59214183      * |     CpG:_36          389
##   [28691]     chrY 59349267-59349574      * |     CpG:_29          308
##                 mid             mid2
##           &lt;numeric&gt;        &lt;numeric&gt;
##       [1]     29273  171.09354166654
##       [2]    135344 367.891288290441
##       [3]    328010  572.72157284321
##       [4]    437658 661.557253758131
##       [5]    449909 670.752562425221
##       ...       ...              ...
##   [28687]  27610602 5254.57914584984
##   [28688]  28555734 5343.75654385564
##   [28689]  28773430 5364.08706118758
##   [28690]  59213989 7695.06263782173
##   [28691]  59349420 7703.85747531715
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>In plyranges and GenomicRanges there are additional helpers that are useful for column creation during data analysis. As an example, we can add the count of the number of times each cpg island overlaps an exon and add a logical column asking if a cpg island is contained entirely within an exon.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(cpg, 
               <span class="dt">n_olap =</span> <span class="kw"><a href="../reference/ranges-count-overlaps.html">count_overlaps</a></span>(cpg, exons),
               <span class="dt">is_contained =</span> cpg <span class="op">%within%</span><span class="st"> </span>exons
               )

cpg2</code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 3 metadata columns:
##           seqnames            ranges strand |        name    n_olap
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116         1
##       [2]     chr1     135125-135563      * |     CpG:_30         1
##       [3]     chr1     327791-328229      * |     CpG:_29         3
##       [4]     chr1     437152-438164      * |     CpG:_84         0
##       [5]     chr1     449274-450544      * |     CpG:_99         0
##       ...      ...               ...    ... .         ...       ...
##   [28687]     chrY 27610116-27611088      * |     CpG:_76         0
##   [28688]     chrY 28555536-28555932      * |     CpG:_32         0
##   [28689]     chrY 28773316-28773544      * |     CpG:_25         0
##   [28690]     chrY 59213795-59214183      * |     CpG:_36         5
##   [28691]     chrY 59349267-59349574      * |     CpG:_29         0
##           is_contained
##              &lt;logical&gt;
##       [1]        FALSE
##       [2]         TRUE
##       [3]         TRUE
##       [4]        FALSE
##       [5]        FALSE
##       ...          ...
##   [28687]        FALSE
##   [28688]        FALSE
##   [28689]        FALSE
##   [28690]        FALSE
##   [28691]        FALSE
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="composing-workflows-with-the-pipe" class="section level2">
<h2 class="hasAnchor">
<a href="#composing-workflows-with-the-pipe" class="anchor"></a>Composing workflows with the pipe <code>%&gt;%</code>
</h2>
<p>Now that you can use <code><a href="../reference/tidyverse-reexports.html">filter()</a></code> and <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>, you might like to start thinking about how you can combine them to form workflows. One way we can do this is with the pipe operator, <code>%&gt;%</code>.</p>
<p>The pipe operator is read as the word <em>then</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># take cpg then filter</span>
cpg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(seqnames <span class="op">==</span><span class="st"> "chrX"</span>)</code></pre></div>
<pre><code>## GRanges object with 896 ranges and 1 metadata column:
##         seqnames              ranges strand |        name
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##     [1]     chrX         64182-64793      * |     CpG:_62
##     [2]     chrX         69134-70029      * |    CpG:_100
##     [3]     chrX       148686-149461      * |     CpG:_85
##     [4]     chrX       166505-167721      * |     CpG:_96
##     [5]     chrX       170415-170686      * |     CpG:_27
##     ...      ...                 ...    ... .         ...
##   [892]     chrX 154687388-154688519      * |    CpG:_131
##   [893]     chrX 154689347-154689788      * |     CpG:_41
##   [894]     chrX 154842113-154842719      * |     CpG:_44
##   [895]     chrX 155110789-155111177      * |     CpG:_36
##   [896]     chrX 155246261-155246568      * |     CpG:_29
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>The argument on the left hand side of the pipe is forwarded to the first argument of the function on the right hand side. This can be made explicit with a <code>.</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(., seqnames <span class="op">==</span><span class="st"> "chrX"</span>)</code></pre></div>
<pre><code>## GRanges object with 896 ranges and 1 metadata column:
##         seqnames              ranges strand |        name
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##     [1]     chrX         64182-64793      * |     CpG:_62
##     [2]     chrX         69134-70029      * |    CpG:_100
##     [3]     chrX       148686-149461      * |     CpG:_85
##     [4]     chrX       166505-167721      * |     CpG:_96
##     [5]     chrX       170415-170686      * |     CpG:_27
##     ...      ...                 ...    ... .         ...
##   [892]     chrX 154687388-154688519      * |    CpG:_131
##   [893]     chrX 154689347-154689788      * |     CpG:_41
##   [894]     chrX 154842113-154842719      * |     CpG:_44
##   [895]     chrX 155110789-155111177      * |     CpG:_36
##   [896]     chrX 155246261-155246568      * |     CpG:_29
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>As long as the the result of applying the pipe returns a GRanges object, multiple functions can be strung together:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(seqnames <span class="op">==</span><span class="st"> "chrX"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(strand <span class="op">==</span><span class="st"> "+"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">new_col =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="kw"><a href="../reference/n.html">n</a></span>())) </code></pre></div>
<pre><code>## GRanges object with 0 ranges and 2 metadata columns:
##    seqnames    ranges strand |        name   new_col
##       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>The use of the <code>.</code> can also make pipelines more succinct, as the argument on the left hand side can be forwarded to multiple parts of the function on the right hand side. We can rewrite, the creation of cpg2 above more compactly as</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg <span class="op">%&gt;%</span><span class="st">  </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">n_olap =</span> <span class="kw"><a href="../reference/ranges-count-overlaps.html">count_overlaps</a></span>(., exons),
         <span class="dt">is_contained =</span> . <span class="op">%within%</span><span class="st"> </span>exons)</code></pre></div>
<pre><code>## GRanges object with 28691 ranges and 3 metadata columns:
##           seqnames            ranges strand |        name    n_olap
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116         1
##       [2]     chr1     135125-135563      * |     CpG:_30         1
##       [3]     chr1     327791-328229      * |     CpG:_29         3
##       [4]     chr1     437152-438164      * |     CpG:_84         0
##       [5]     chr1     449274-450544      * |     CpG:_99         0
##       ...      ...               ...    ... .         ...       ...
##   [28687]     chrY 27610116-27611088      * |     CpG:_76         0
##   [28688]     chrY 28555536-28555932      * |     CpG:_32         0
##   [28689]     chrY 28773316-28773544      * |     CpG:_25         0
##   [28690]     chrY 59213795-59214183      * |     CpG:_36         5
##   [28691]     chrY 59349267-59349574      * |     CpG:_29         0
##           is_contained
##              &lt;logical&gt;
##       [1]        FALSE
##       [2]         TRUE
##       [3]         TRUE
##       [4]        FALSE
##       [5]        FALSE
##       ...          ...
##   [28687]        FALSE
##   [28688]        FALSE
##   [28689]        FALSE
##   [28690]        FALSE
##   [28691]        FALSE
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<div id="excercises" class="section level3">
<h3 class="hasAnchor">
<a href="#excercises" class="anchor"></a>Excercises</h3>
<ol style="list-style-type: decimal">
<li>Modify the above example so only CpG islands that are completely within exons are included in the result.</li>
<li>Create a new column in exons called tx_id (the transcript id is everything before _exon in name).</li>
</ol>
</div>
</div>
<div id="modifying-genomic-regions-with-mutate" class="section level2">
<h2 class="hasAnchor">
<a href="#modifying-genomic-regions-with-mutate" class="anchor"></a>Modifying genomic regions with <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>
</h2>
<p>We have learnt in the above sections how to add new metadata columns with <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code> but we can also use `mutate() to modify the genomic regions. Recall that there are three components that make up a genomic region; the start, end, and the width.</p>
<p>Let’s go back to our original GRanges of yeast genes. What happens if we add 10 bases to the width?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 3 metadata columns:
##       seqnames    ranges strand |     gene_id                gc     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;numeric&gt; &lt;integer&gt;
##   [1]       VI 3322-3846      - |     YFL064C 0.406736710341647         0
##   [2]       VI 3030-3338      - |     YFL065C 0.449663328705356         1
##   [3]       VI 1437-2615      - |     YFL066C 0.868391741998494         0
##   [4]       VI 5066-5521      + |     YFL063W 0.935484018875286         0
##   [5]       VI 6426-7565      + |     YFL062W 0.896344820735976         0
##   [6]       VI  836-1363      + |     YFL067W 0.765752661740407         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">width =</span> width <span class="op">+</span><span class="st"> </span>10L)</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 3 metadata columns:
##       seqnames    ranges strand |     gene_id                gc     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;numeric&gt; &lt;integer&gt;
##   [1]       VI 3322-3856      - |     YFL064C 0.406736710341647         0
##   [2]       VI 3030-3348      - |     YFL065C 0.449663328705356         1
##   [3]       VI 1437-2625      - |     YFL066C 0.868391741998494         0
##   [4]       VI 5066-5531      + |     YFL063W 0.935484018875286         0
##   [5]       VI 6426-7575      + |     YFL062W 0.896344820735976         0
##   [6]       VI  836-1373      + |     YFL067W 0.765752661740407         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>All three components are mutally dependent, so in order to preserve the integrity of the GRanges structure, either the start or the end positions must change. In the case of adding to the width position by default the start is fixed and the end is increased. But what if we would like to modify the width but leave the end of fixed? Or the midpoint fixed?</p>
<p>The solution is to anchor those positions (i.e. leave them fixed), using the <code>anchor_*()</code> family of functions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># fix the end</span>
genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-anchor.html">anchor_end</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">width =</span> width <span class="op">+</span><span class="st"> </span>10L)</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 3 metadata columns:
##       seqnames    ranges strand |     gene_id                gc     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;numeric&gt; &lt;integer&gt;
##   [1]       VI 3312-3846      - |     YFL064C 0.406736710341647         0
##   [2]       VI 3020-3338      - |     YFL065C 0.449663328705356         1
##   [3]       VI 1427-2615      - |     YFL066C 0.868391741998494         0
##   [4]       VI 5056-5521      + |     YFL063W 0.935484018875286         0
##   [5]       VI 6416-7565      + |     YFL062W 0.896344820735976         0
##   [6]       VI  826-1363      + |     YFL067W 0.765752661740407         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># fix the centre</span>
genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-anchor.html">anchor_centre</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">width =</span> width <span class="op">+</span><span class="st"> </span>10L)</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 3 metadata columns:
##       seqnames    ranges strand |     gene_id                gc     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;numeric&gt; &lt;integer&gt;
##   [1]       VI 3317-3851      - |     YFL064C 0.406736710341647         0
##   [2]       VI 3025-3343      - |     YFL065C 0.449663328705356         1
##   [3]       VI 1432-2620      - |     YFL066C 0.868391741998494         0
##   [4]       VI 5061-5526      + |     YFL063W 0.935484018875286         0
##   [5]       VI 6421-7570      + |     YFL062W 0.896344820735976         0
##   [6]       VI  831-1368      + |     YFL067W 0.765752661740407         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>There are additional anchors that account for the strandedness of features, we can fix the 3’ or 5’ parts of a GRanges object using <code><a href="../reference/ranges-anchor.html">anchor_3p()</a></code> or <code><a href="../reference/ranges-anchor.html">anchor_5p()</a></code>, respectively.</p>
<p>The plyranges and GenomicRanges API also have utilities to modify genomic regions via standard set operations like taking the complement (<code><a href="../reference/ranges-setops.html">complement_ranges()</a></code>) which will find gaps, or finding flanking ranges (see <code><a href="../reference/flank-ranges.html">flank_left()</a></code> and friends) or promoters (<code><a href="https://www.rdocumentation.org/packages/GenomicRanges/topics/intra-range-methods">GenomicRanges::promoters()</a></code>).</p>
<div id="exercises" class="section level3">
<h3 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h3>
<ol style="list-style-type: decimal">
<li>Create a new GRanges object from CpG islands that stretches their ranges by their width while leaving their mid point fixed.</li>
<li>Create a new GRanges object from exons that has only non-exonic regions.</li>
<li>Create a new GRanges object that represent 2bp canonical splice sites on either side of exon.</li>
</ol>
</div>
</div>
<div id="summarise-annoations-and-measurements-with-summarise" class="section level2">
<h2 class="hasAnchor">
<a href="#summarise-annoations-and-measurements-with-summarise" class="anchor"></a>Summarise annoations and measurements with <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code>
</h2>
<p>We can compute summarise columns GRanges object using <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code>. Using <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code> is very similiar to <code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>, we construct new columns that are summaries of the columns in the input the GRanges.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(
    <span class="dt">count =</span> <span class="kw"><a href="../reference/n.html">n</a></span>(),
    <span class="dt">median_width =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/median">median</a></span>(width)
  )</code></pre></div>
<pre><code>## DataFrame with 1 row and 2 columns
##       count median_width
##   &lt;integer&gt;    &lt;numeric&gt;
## 1    459752          133</code></pre>
<p>The main difference is that by summarising our data, we lose the necessary information required to form a valid GRanges. Instead, <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code> returns a <code>DataFrame</code> (see <code><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">S4Vectors::DataFrame()</a></code> for more information). The <code>DataFrame</code> is the Bioconductor version of the base R data.frame. It works mostly the same way as the base R <code>data.frame</code> but can also hold data structures that are common in Bioconductor (like <code>Rle</code> mentioned above).</p>
<div id="exercises-1" class="section level3">
<h3 class="hasAnchor">
<a href="#exercises-1" class="anchor"></a>Exercises</h3>
<ol style="list-style-type: decimal">
<li>What is the average length of CpG islands accross the genome?</li>
<li>How many distinct transcript identifers are there in exons?</li>
</ol>
</div>
</div>
<div id="summarise-over-the-genome-with-reduce_ranges-and-disjoin_ranges" class="section level2">
<h2 class="hasAnchor">
<a href="#summarise-over-the-genome-with-reduce_ranges-and-disjoin_ranges" class="anchor"></a>Summarise over the genome with <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code> and <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code>
</h2>
<p>To preserve the GRanges data structure while performing summaries, it is often useful to either merge overlapping ranges (called reducing) or expand ranges by finding the union of the end points or ranges (called disjoining).</p>
<p>To make this more concrete let’s <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code> and <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code> on the <code>genes</code> object we defined above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># to make things easier to compare </span>
<span class="co"># sort by the start and end coordinates</span>
genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">arrange</a></span>(start, end)</code></pre></div>
<pre><code>## GRanges object with 6 ranges and 3 metadata columns:
##       seqnames    ranges strand |     gene_id                gc     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;numeric&gt; &lt;integer&gt;
##   [1]       VI  836-1363      + |     YFL067W 0.765752661740407         0
##   [2]       VI 1437-2615      - |     YFL066C 0.868391741998494         0
##   [3]       VI 3030-3338      - |     YFL065C 0.449663328705356         1
##   [4]       VI 3322-3846      - |     YFL064C 0.406736710341647         0
##   [5]       VI 5066-5521      + |     YFL063W 0.935484018875286         0
##   [6]       VI 6426-7565      + |     YFL062W 0.896344820735976         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reduce_ranges will merge the 3rd and 4th ranges above</span>
genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-reduce.html">reduce_ranges</a></span>()</code></pre></div>
<pre><code>## GRanges object with 5 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]       VI  836-1363      *
##   [2]       VI 1437-2615      *
##   [3]       VI 3030-3846      *
##   [4]       VI 5066-5521      *
##   [5]       VI 6426-7565      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># disjoin ranges will split the 3rd and 4th ranges into</span>
<span class="co"># 3 disjoint ranges. </span>
genes <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-disjoin.html">disjoin_ranges</a></span>()</code></pre></div>
<pre><code>## GRanges object with 7 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]       VI  836-1363      *
##   [2]       VI 1437-2615      *
##   [3]       VI 3030-3321      *
##   [4]       VI 3322-3338      *
##   [5]       VI 3339-3846      *
##   [6]       VI 5066-5521      *
##   [7]       VI 6426-7565      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Note that by default, <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code> and <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code> will discard any strand information. To preserve them use <code><a href="../reference/ranges-reduce.html">reduce_ranges_directed()</a></code> and <code><a href="../reference/ranges-disjoin.html">disjoin_ranges_directed()</a></code>.</p>
<p>The <code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code> and <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code> can perform summarisation too. For example, we can reduce exons and count the number of exons that were merged together, and concatante the names of the exons that were merged.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-reduce.html">reduce_ranges</a></span>(
    <span class="dt">count =</span> <span class="kw"><a href="../reference/n.html">n</a></span>(),
    <span class="dt">name =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(name, <span class="dt">collapse =</span> <span class="st">","</span>)
  )</code></pre></div>
<pre><code>## GRanges object with 229241 ranges and 2 metadata columns:
##                  seqnames        ranges strand |     count
##                     &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1   11874-12227      * |         1
##        [2]           chr1   12613-12721      * |         1
##        [3]           chr1   13221-14829      * |         2
##        [4]           chr1   14970-15038      * |         1
##        [5]           chr1   15796-15947      * |         1
##        ...            ...           ...    ... .       ...
##   [229237] chrUn_gl000228 105896-107275      * |         6
##   [229238] chrUn_gl000228 109202-110581      * |         6
##   [229239] chrUn_gl000228 112508-113887      * |         7
##   [229240] chrUn_gl000228 114024-114115      * |         1
##   [229241] chrUn_gl000228 114478-114676      * |         1
##                                                                                                                                                                                                                                                                                                                                   name
##                                                                                                                                                                                                                                                                                                                            &lt;character&gt;
##        [1]                                                                                                                                                                                                                                                                                             NR_046018_exon_0_0_chr1_11874_f
##        [2]                                                                                                                                                                                                                                                                                             NR_046018_exon_1_0_chr1_12613_f
##        [3]                                                                                                                                                                                                                                                             NR_046018_exon_2_0_chr1_13221_f,NR_024540_exon_0_0_chr1_14362_r
##        [4]                                                                                                                                                                                                                                                                                             NR_024540_exon_1_0_chr1_14970_r
##        [5]                                                                                                                                                                                                                                                                                             NR_024540_exon_2_0_chr1_15796_r
##        ...                                                                                                                                                                                                                                                                                                                         ...
##   [229237]                                            NM_001127386_exon_0_0_chrUn_gl000228_105896_f,NM_001127388_exon_0_0_chrUn_gl000228_105896_f,NM_001127389_exon_0_0_chrUn_gl000228_105896_f,NM_001164467_exon_0_0_chrUn_gl000228_105896_f,NM_001278056_exon_0_0_chrUn_gl000228_105896_f,NM_033178_exon_0_0_chrUn_gl000228_105896_f
##   [229238]                                            NM_001127386_exon_0_0_chrUn_gl000228_109202_f,NM_001127388_exon_0_0_chrUn_gl000228_109202_f,NM_001127389_exon_0_0_chrUn_gl000228_109202_f,NM_001164467_exon_0_0_chrUn_gl000228_109202_f,NM_001278056_exon_0_0_chrUn_gl000228_109202_f,NM_033178_exon_0_0_chrUn_gl000228_109202_f
##   [229239] NM_001127386_exon_0_0_chrUn_gl000228_112508_f,NM_001127388_exon_0_0_chrUn_gl000228_112508_f,NM_001127389_exon_0_0_chrUn_gl000228_112508_f,NM_001164467_exon_0_0_chrUn_gl000228_112508_f,NM_001278056_exon_0_0_chrUn_gl000228_112508_f,NM_033178_exon_0_0_chrUn_gl000228_112508_f,NR_038191_exon_0_0_chrUn_gl000228_112605_f
##   [229240]                                                                                                                                                                                                                                                                                  NR_038191_exon_1_0_chrUn_gl000228_114024_f
##   [229241]                                                                                                                                                                                                                                                                                  NR_038191_exon_2_0_chrUn_gl000228_114478_f
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<div id="excerises" class="section level3">
<h3 class="hasAnchor">
<a href="#excerises" class="anchor"></a>Excerises</h3>
<ol style="list-style-type: decimal">
<li>Reduce CpG islands, what is the range that has the maximum number of nearby features?</li>
<li>Confirm your results in 1. using <code><a href="../reference/ranges-filter-overlaps.html">filter_by_overlaps()</a></code>.</li>
</ol>
</div>
</div>
<div id="analyses-over-groups-with-group_by" class="section level2">
<h2 class="hasAnchor">
<a href="#analyses-over-groups-with-group_by" class="anchor"></a>Analyses over groups with <code><a href="../reference/tidyverse-reexports.html">group_by()</a></code>
</h2>
<p>Any of the above operations can be perfomed over subsets of the <code>GRanges</code> object with <code><a href="../reference/tidyverse-reexports.html">group_by()</a></code>, this often useful if you want to <code><a href="../reference/tidyverse-reexports.html">filter()</a></code>/<code><a href="../reference/tidyverse-reexports.html">mutate()</a></code>/<code><a href="../reference/tidyverse-reexports.html">summarise()</a></code>/<code><a href="../reference/ranges-reduce.html">reduce_ranges()</a></code>/<code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code> over annotation identifiers or chromosomes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(seqnames) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(
    <span class="dt">count =</span> <span class="kw"><a href="../reference/n.html">n</a></span>(),
    <span class="dt">median_width =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/median">median</a></span>(width)
  )</code></pre></div>
<pre><code>## DataFrame with 49 rows and 3 columns
##           seqnames     count median_width
##              &lt;Rle&gt; &lt;integer&gt;    &lt;numeric&gt;
## 1             chr1     43366          137
## 2             chr2     34022          132
## 3             chr3     25183          135
## 4             chr4     16265          135
## 5             chr5     18229          133
## ...            ...       ...          ...
## 45  chrUn_gl000219         7          111
## 46  chrUn_gl000220        10          263
## 47  chrUn_gl000222        20          160
## 48  chrUn_gl000223        22        179.5
## 49  chrUn_gl000228        85         1380</code></pre>
<div id="excercises-1" class="section level3">
<h3 class="hasAnchor">
<a href="#excercises-1" class="anchor"></a>Excercises</h3>
<p>Let’s read in another BED file that has functional annotations output by chromHMM from ENCODE human embrionic stem cell ChIP-seq data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hesc_bed &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"hesc.chromHmm.bed"</span>, 
                        <span class="dt">package=</span><span class="st">"HelloRangesData"</span>)

hesc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/io-bed-read.html">read_bed</a></span>(hesc_bed, <span class="dt">genome_info =</span> build)
hesc</code></pre></div>
<pre><code>## GRanges object with 619061 ranges and 1 metadata column:
##            seqnames              ranges strand |              name
##               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |       &lt;character&gt;
##        [1]     chr1         10001-10600      * | 15_Repetitive/CNV
##        [2]     chr1         10601-11137      * | 13_Heterochrom/lo
##        [3]     chr1         11138-11537      * |       8_Insulator
##        [4]     chr1         11538-11937      * |       11_Weak_Txn
##        [5]     chr1         11938-12137      * | 14_Repetitive/CNV
##        ...      ...                 ...    ... .               ...
##   [619057]     chrX 155256607-155258006      * |       11_Weak_Txn
##   [619058]     chrX 155258007-155258606      * |       8_Insulator
##   [619059]     chrX 155258607-155259606      * |       11_Weak_Txn
##   [619060]     chrX 155259607-155259806      * | 14_Repetitive/CNV
##   [619061]     chrX 155259807-155260406      * | 15_Repetitive/CNV
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>The name column is the predicted state of a range from chromHMM.</p>
<p>Use the hesc GRanges object to answer the following:</p>
<ol style="list-style-type: decimal">
<li>How many ranges are there for each predicted state?</li>
<li>How many base pairs are represented in the genome for each predicted state? Which state has the maximum number of bases?</li>
<li>How many base pairs are represented in the genome by all predicted states?</li>
</ol>
</div>
</div>
<div id="finding-overlaps-with-the-join-operator" class="section level2">
<h2 class="hasAnchor">
<a href="#finding-overlaps-with-the-join-operator" class="anchor"></a>Finding overlaps with the join operator</h2>
<p>A common operation on two GRanges is to find overlaps and to keep metadata associated with both GRanges objects when an overlap has occurred. We refer to this operation as an <em>overlap join</em>. There are three types of overlap join: inner (A), intersect (B), and left (C) that are shown on the figure below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/include_graphics">include_graphics</a></span>(<span class="st">"olaps.png"</span>)</code></pre></div>
<p><img src="olaps.png" width="\textwidth"></p>
<p>The first type of overlap join is called <code><a href="../reference/overlap-joins.html">join_overlap_inner()</a></code>, we could use this to find which CpG islands overlap exons</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg2  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(cpg, exons)
cpg2</code></pre></div>
<pre><code>## GRanges object with 45500 ranges and 3 metadata columns:
##           seqnames            ranges strand |      name.x
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116
##       [2]     chr1     135125-135563      * |     CpG:_30
##       [3]     chr1     327791-328229      * |     CpG:_29
##       [4]     chr1     327791-328229      * |     CpG:_29
##       [5]     chr1     327791-328229      * |     CpG:_29
##       ...      ...               ...    ... .         ...
##   [45496]     chrY 59213795-59214183      * |     CpG:_36
##   [45497]     chrY 59213795-59214183      * |     CpG:_36
##   [45498]     chrY 59213795-59214183      * |     CpG:_36
##   [45499]     chrY 59213795-59214183      * |     CpG:_36
##   [45500]     chrY 59213795-59214183      * |     CpG:_36
##                                          name.y     score
##                                     &lt;character&gt; &lt;numeric&gt;
##       [1]      NR_024540_exon_10_0_chr1_29321_r         0
##       [2]      NR_039983_exon_0_0_chr1_134773_r         0
##       [3]      NR_028322_exon_2_0_chr1_324439_f         0
##       [4]      NR_028325_exon_2_0_chr1_324439_f         0
##       [5]      NR_028327_exon_3_0_chr1_327036_f         0
##       ...                                   ...       ...
##   [45496] NM_001145149_exon_0_0_chrY_59213949_f         0
##   [45497] NM_001185183_exon_0_0_chrY_59213949_f         0
##   [45498]    NM_005638_exon_0_0_chrY_59213949_f         0
##   [45499]    NR_033714_exon_0_0_chrY_59213949_f         0
##   [45500]    NR_033715_exon_0_0_chrY_59213949_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Often we would are also interested in the amount of overlap, we can use the second type of overlap join <code><a href="../reference/overlap-joins.html">join_overlap_intersect()</a></code> to do this. We can modify the above example, to find which CpG islands overlap exons and how much the CpG island ranges overlap the exon ranges.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_intersect</a></span>(cpg, exons)
cpg3</code></pre></div>
<pre><code>## GRanges object with 45500 ranges and 3 metadata columns:
##           seqnames            ranges strand |      name.x
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]     chr1       29321-29370      * |    CpG:_116
##       [2]     chr1     135125-135563      * |     CpG:_30
##       [3]     chr1     327791-328229      * |     CpG:_29
##       [4]     chr1     327791-328229      * |     CpG:_29
##       [5]     chr1     327791-328229      * |     CpG:_29
##       ...      ...               ...    ... .         ...
##   [45496]     chrY 59213949-59214117      * |     CpG:_36
##   [45497]     chrY 59213949-59214117      * |     CpG:_36
##   [45498]     chrY 59213949-59214117      * |     CpG:_36
##   [45499]     chrY 59213949-59214117      * |     CpG:_36
##   [45500]     chrY 59213949-59214117      * |     CpG:_36
##                                          name.y     score
##                                     &lt;character&gt; &lt;numeric&gt;
##       [1]      NR_024540_exon_10_0_chr1_29321_r         0
##       [2]      NR_039983_exon_0_0_chr1_134773_r         0
##       [3]      NR_028322_exon_2_0_chr1_324439_f         0
##       [4]      NR_028325_exon_2_0_chr1_324439_f         0
##       [5]      NR_028327_exon_3_0_chr1_327036_f         0
##       ...                                   ...       ...
##   [45496] NM_001145149_exon_0_0_chrY_59213949_f         0
##   [45497] NM_001185183_exon_0_0_chrY_59213949_f         0
##   [45498]    NM_005638_exon_0_0_chrY_59213949_f         0
##   [45499]    NR_033714_exon_0_0_chrY_59213949_f         0
##   [45500]    NR_033715_exon_0_0_chrY_59213949_f         0
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>Finally, we sometimes would like to know where overlap occurs but also return any other ranges that are not overlapping. This is this the final type of overlap join <code><a href="../reference/overlap-joins.html">join_overlap_left()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_left</a></span>(cpg, exons)
cpg4</code></pre></div>
<pre><code>## GRanges object with 55346 ranges and 3 metadata columns:
##           seqnames            ranges strand |      name.x
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       [1]     chr1       28736-29810      * |    CpG:_116
##       [2]     chr1     135125-135563      * |     CpG:_30
##       [3]     chr1     327791-328229      * |     CpG:_29
##       [4]     chr1     327791-328229      * |     CpG:_29
##       [5]     chr1     327791-328229      * |     CpG:_29
##       ...      ...               ...    ... .         ...
##   [55342]     chrY 59213795-59214183      * |     CpG:_36
##   [55343]     chrY 59213795-59214183      * |     CpG:_36
##   [55344]     chrY 59213795-59214183      * |     CpG:_36
##   [55345]     chrY 59213795-59214183      * |     CpG:_36
##   [55346]     chrY 59349267-59349574      * |     CpG:_29
##                                          name.y     score
##                                     &lt;character&gt; &lt;numeric&gt;
##       [1]      NR_024540_exon_10_0_chr1_29321_r         0
##       [2]      NR_039983_exon_0_0_chr1_134773_r         0
##       [3]      NR_028322_exon_2_0_chr1_324439_f         0
##       [4]      NR_028325_exon_2_0_chr1_324439_f         0
##       [5]      NR_028327_exon_3_0_chr1_327036_f         0
##       ...                                   ...       ...
##   [55342] NM_001185183_exon_0_0_chrY_59213949_f         0
##   [55343]    NM_005638_exon_0_0_chrY_59213949_f         0
##   [55344]    NR_033714_exon_0_0_chrY_59213949_f         0
##   [55345]    NR_033715_exon_0_0_chrY_59213949_f         0
##   [55346]                                  &lt;NA&gt;      &lt;NA&gt;
##   -------
##   seqinfo: 94 sequences (1 circular) from 2 genomes (NA, hg19)</code></pre>
<p>The type of overlap can be modified, by adding suffixes to the function call for example, to find overlaps that are within we could use <code><a href="../reference/overlap-joins.html">join_overlap_inner_within()</a></code>.</p>
<div id="exercies" class="section level3">
<h3 class="hasAnchor">
<a href="#exercies" class="anchor"></a>Exercies</h3>
<ol style="list-style-type: decimal">
<li><p>Create a new GRanges object, that has exons that are completely within Enhancers elements of hesc. How many exons are there?</p></li>
<li><p>Use <code><a href="../reference/overlap-joins.html">join_overlap_intersect()</a></code> to filter exons if at least 50 per cent of their bases are overlapped by enhancer elements.</p></li>
<li><p>Count the number of each enchancer element type that is exonic. There are several ways of doing this, but see if you can come up with a solution using <code><a href="../reference/overlap-joins.html">join_overlap_left()</a></code> + <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</p></li>
</ol>
</div>
</div>
<div id="summarising-coverage-with-compute_coverage" class="section level2">
<h2 class="hasAnchor">
<a href="#summarising-coverage-with-compute_coverage" class="anchor"></a>Summarising coverage with <code><a href="../reference/compute_coverage.html">compute_coverage()</a></code>
</h2>
<p>Often we are interested in counting the number of features over the genome that overlap each other. In <code>plyranges</code> this is done with the <code><a href="../reference/compute_coverage.html">compute_coverage()</a></code> function in combination with other functions to manipulate the results. This function will always return a new GRanges with a column called score which contains the coverage value over a range.</p>
<div id="coverage-values" class="section level3">
<h3 class="hasAnchor">
<a href="#coverage-values" class="anchor"></a>Coverage values</h3>
<p>By default, <code><a href="../reference/compute_coverage.html">compute_coverage()</a></code> returns all coverage scores even if there is no coverage over a genomic region.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cvg &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/compute_coverage.html">compute_coverage</a></span>()

cvg</code></pre></div>
<pre><code>## GRanges object with 475194 ranges and 1 metadata column:
##                  seqnames      ranges strand |     score
##                     &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1     1-11873      * |         0
##        [2]           chr1 11874-12227      * |         1
##        [3]           chr1 12228-12612      * |         0
##        [4]           chr1 12613-12721      * |         1
##        [5]           chr1 12722-13220      * |         0
##        ...            ...         ...    ... .       ...
##   [475190] chrUn_gl000245     1-36651      * |         0
##   [475191] chrUn_gl000246     1-38154      * |         0
##   [475192] chrUn_gl000247     1-36422      * |         0
##   [475193] chrUn_gl000248     1-39786      * |         0
##   [475194] chrUn_gl000249     1-38502      * |         0
##   -------
##   seqinfo: 93 sequences from an unspecified genome</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cvg <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(score <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</code></pre></div>
<pre><code>## GRanges object with 245860 ranges and 1 metadata column:
##                  seqnames        ranges strand |     score
##                     &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1   11874-12227      * |         1
##        [2]           chr1   12613-12721      * |         1
##        [3]           chr1   13221-14361      * |         1
##        [4]           chr1   14362-14409      * |         2
##        [5]           chr1   14410-14829      * |         1
##        ...            ...           ...    ... .       ...
##   [245856] chrUn_gl000228 109202-110581      * |         6
##   [245857] chrUn_gl000228 112508-112604      * |         6
##   [245858] chrUn_gl000228 112605-113887      * |         7
##   [245859] chrUn_gl000228 114024-114115      * |         1
##   [245860] chrUn_gl000228 114478-114676      * |         1
##   -------
##   seqinfo: 93 sequences from an unspecified genome</code></pre>
<p>Also note that computing coverage will result in a GRanges that is disjoint, which you can check with <code><a href="https://www.rdocumentation.org/packages/GenomicRanges/topics/inter-range-methods">isDisjoint()</a></code> from GenomicRanges.</p>
</div>
<div id="coverage-histogram" class="section level3">
<h3 class="hasAnchor">
<a href="#coverage-histogram" class="anchor"></a>Coverage histogram</h3>
<p>We can <code><a href="../reference/tidyverse-reexports.html">summarise()</a></code> the coverage scores as a histogram over all chromosomes in a genome using <code><a href="../reference/tidyverse-reexports.html">group_by()</a></code>.</p>
<p>We count the total number of bases (that is, the width of the range) over each chromosome over each coverage score.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cvg_hist &lt;-<span class="st"> </span>cvg <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(seqnames, score) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(<span class="dt">count =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(width))</code></pre></div>
<p>We can also compute the coverage histogram over the entire genome by grouping over the scores. Note that we coerce the width to a numeric to avoid integer overflow, when summing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cvg_hist_all &lt;-<span class="st"> </span>cvg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">total_length =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(width)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(score) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(
    <span class="dt">seqnames =</span> <span class="st">"genome"</span>, 
    <span class="dt">count =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(total_length)
  )</code></pre></div>
<p>Then the results can be put togeher using <code><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind()</a></code> or plotted using your favourite plotting pacakge.</p>
</div>
<div id="a-more-complicated-example" class="section level3">
<h3 class="hasAnchor">
<a href="#a-more-complicated-example" class="anchor"></a>A more complicated example</h3>
<p>Here we perform another example where we find exons with no coverage that overlap cpg islands. To compose a chain first we compute the coverage then use a filter to restrict to zero coverage ranges. Then we perfrom an inner overlap join to find which cpg islands overlap the exonic regions with zero coverage.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">overlaps &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/compute_coverage.html">compute_coverage</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(score <span class="op">==</span><span class="st"> </span>0L) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(.,  cpg) 

overlaps</code></pre></div>
<pre><code>## GRanges object with 40913 ranges and 2 metadata columns:
##                 seqnames        ranges strand |     score        name
##                    &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##       [1]           chr1   24892-29320      * |         0    CpG:_116
##       [2]           chr1   29371-34610      * |         0    CpG:_116
##       [3]           chr1 368598-621095      * |         0     CpG:_84
##       [4]           chr1 368598-621095      * |         0     CpG:_99
##       [5]           chr1 368598-621095      * |         0     CpG:_94
##       ...            ...           ...    ... .       ...         ...
##   [40909] chrUn_gl000237       1-45867      * |         0     CpG:_30
##   [40910] chrUn_gl000240       1-41933      * |         0     CpG:_17
##   [40911] chrUn_gl000241       1-42152      * |         0     CpG:_71
##   [40912] chrUn_gl000242       1-43523      * |         0     CpG:_20
##   [40913] chrUn_gl000243       1-43341      * |         0     CpG:_31
##   -------
##   seqinfo: 93 sequences from an unspecified genome</code></pre>
<p>We can also compute the coverage histogram of exons over cpg islands and then plot results as an ecdf.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cvg_over_exons &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/compute_coverage.html">compute_coverage</a></span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(., cpg) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(
    <span class="dt">total_length =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(width),
    <span class="dt">genome_length =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(total_length)
  ) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(score, genome_length) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(
    <span class="dt">count =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(total_length),
    <span class="dt">fraction =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(total_length) <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(genome_length)
  ) 

cvg_over_exons</code></pre></div>
<pre><code>## DataFrame with 28 rows and 4 columns
##         score genome_length      count             fraction
##     &lt;integer&gt;     &lt;numeric&gt;  &lt;numeric&gt;        &lt;NumericList&gt;
## 1           0    3288637798 3275850528     0.99611168186178
## 2           1    3288637798    8565943   0.0026047085529484
## 3           2    3288637798    2429912 0.000738881004614665
## 4           3    3288637798     915394 0.000278350507482673
## 5           4    3288637798     373081 0.000113445451556535
## ...       ...           ...        ...                  ...
## 24         25    3288637798        161 4.89564402920604e-08
## 25         30    3288637798         21 6.38562264679049e-09
## 26         31    3288637798        160 4.86523630231656e-08
## 27         33    3288637798        979 2.97691646247995e-07
## 28         77    3288637798       1157 3.51817400111266e-07</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cumsum">cumsum</a></span>(fraction) <span class="op">~</span><span class="st"> </span>score, 
     <span class="dt">data =</span> cvg_over_exons, 
     <span class="dt">xlim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">30</span>), 
     <span class="dt">type=</span> <span class="st">"s"</span>, 
     <span class="dt">xlab =</span> <span class="st">"coverage"</span>, 
     <span class="dt">ylab =</span> <span class="st">"fraction of bp &gt; coverage"</span>)</code></pre></div>
<p><img src="common-tasks_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
</div>
<div id="learning-more" class="section level1">
<h1 class="hasAnchor">
<a href="#learning-more" class="anchor"></a>Learning more</h1>
<p>We have touched the surface of what’s possible with plyranges and GenomicRanges. A good starting point to learn more about GenomicRanges is <a href="https://bioconductor.github.io/BiocWorkshops/solving-common-bioinformatic-challenges-using-genomicranges.html">here</a> and plyranges is <a href="https://bioconductor.github.io/BiocWorkshops/fluent-genomic-data-analysis-with-plyranges.html">here</a>.</p>
<p>If you have any questions about the use of these packages, please ask a question on <a href="https://support.bioconductor.org/">support.bioconductor.org</a> or post an issue on <a href="https://github.com/sa-lee/plyranges">github</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#who-is-this-tutorial-is-for">Who is this tutorial is for?</a></li>
      <li><a href="#setup">Setup</a></li>
      </ul>
</li>
      <li><a href="#introducing-granges">Introducing GRanges</a></li>
      <li>
<a href="#granges-transformation-and-import">GRanges transformation and import</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting started</a></li>
      <li><a href="#filter-ranges-with-filter">Filter ranges with <code><a href="--/reference/tidyverse-reexports-html">filter()</a></code></a></li>
      <li><a href="#modify-metadata-with-mutate">Modify metadata with <code><a href="--/reference/tidyverse-reexports-html">mutate()</a></code></a></li>
      <li><a href="#composing-workflows-with-the-pipe">Composing workflows with the pipe <code>%&gt;%</code></a></li>
      <li><a href="#modifying-genomic-regions-with-mutate">Modifying genomic regions with <code><a href="--/reference/tidyverse-reexports-html">mutate()</a></code></a></li>
      <li><a href="#summarise-annoations-and-measurements-with-summarise">Summarise annoations and measurements with <code><a href="--/reference/tidyverse-reexports-html">summarise()</a></code></a></li>
      <li><a href="#summarise-over-the-genome-with-reduce_ranges-and-disjoin_ranges">Summarise over the genome with <code><a href="--/reference/ranges-reduce-html">reduce_ranges()</a></code> and <code><a href="--/reference/ranges-disjoin-html">disjoin_ranges()</a></code></a></li>
      <li><a href="#analyses-over-groups-with-group_by">Analyses over groups with <code><a href="--/reference/tidyverse-reexports-html">group_by()</a></code></a></li>
      <li><a href="#finding-overlaps-with-the-join-operator">Finding overlaps with the join operator</a></li>
      <li><a href="#summarising-coverage-with-compute_coverage">Summarising coverage with <code><a href="--/reference/compute_coverage-html">compute_coverage()</a></code></a></li>
      </ul>
</li>
      <li><a href="#learning-more">Learning more</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://stuartlee.org">Stuart Lee</a>, <a href="https://github.com/lawremi">Michael Lawrence</a>, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
