<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to the plyranges package â€¢ plyranges</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">plyranges</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bedtools-examples.html">Using plyranges to perform common genomic data wrangling tasks</a>
    </li>
    <li>
      <a href="../articles/introduction-plyranges.html">Introduction to the plyranges package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Introduction to the plyranges package</h1>
                        <h4 class="author">Stuart Lee</h4>
            
            <h4 class="date">2017-11-23</h4>
          </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      A fluent interface to <code>Ranges</code> classes.
    </div>
    
<div class="contents">
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <code>plyranges</code> package provides a <a href="https://en.wikipedia.org/wiki/Fluent_interface">fluent</a> interface for performing common genomic data analysis tasks. One goal of this package is to decrease the learning curve for Bioconductor and S4 classes (especially for new users) by providing a consistent interface to common classes, <code>IRanges</code> and <code>GRanges</code> classes.</p>
<p>This document provides an overview of some of the features of the <code>plyranges</code> grammar for analysing genomic data. I give a brief introduction to the <code>Ranges</code> and <code>GenomicRanges</code> classes and then detail each feature of the <code>plyranges</code> library: construction, arithmetic, restriction, summarisation, joins, and data import/output.</p>
</div>
<div id="ranges-revisted" class="section level1">
<h1 class="hasAnchor">
<a href="#ranges-revisted" class="anchor"></a>Ranges revisted</h1>
<p><code>Ranges</code> objects can either represent intervals of integers as <code>IRanges</code> (which have start, end and width attributes) or represent genomic intervals (which have additional attributes, sequence name, and strand) as <code>GRanges</code>. In addition, both types of <code>Ranges</code> can store information about their intervals as metadata columns (for example GC content over a genomic interval).</p>
<p><code>Ranges</code> objects follow the tidy data principle: each row of a <code>Ranges</code> object corresponds to an interval, while each column will represent a variable about that interval, and generally each object will represent a single unit of observation (like gene annotations).</p>
<p>Consequently, <code>Ranges</code> objects provide a powerful representation for reasoning about genomic data. In this vignette, you will learn more about <code>Ranges</code> objects and how via grouping, restriction and summarisation you can perform common data tasks.</p>
</div>
<div id="constructing-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#constructing-ranges" class="anchor"></a>Constructing Ranges</h1>
<p>To construct an <code>IRanges</code> we require there are at least two columns that represent at either a starting coordinate, finishing coordinate or the width of the interval. To construct a <code>GRanges</code> we require a column that represents that sequence name ( contig or chromosome id), and an optional column to represent the strandedness of an interval.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(plyranges))
<span class="kw">set.seed</span>(<span class="dv">100</span>)
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">2</span><span class="op">:-</span><span class="dv">1</span>, <span class="dv">13</span><span class="op">:</span><span class="dv">15</span>), 
                 <span class="dt">width=</span><span class="kw">c</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">2</span><span class="op">:</span><span class="dv">0</span>))

<span class="co"># produces IRanges</span>
rng &lt;-<span class="st"> </span>df <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>()
rng</code></pre></div>
<pre><code>## IRanges object with 7 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         1         0
##   [2]         1         1         1
##   [3]         0         1         2
##   [4]        -1         1         3
##   [5]        13        14         2
##   [6]        14        14         1
##   [7]        15        14         0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># seqname is required for GRanges, metadata is automatically kept</span>
grng &lt;-<span class="st"> </span>df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">transform</span>(<span class="dt">seqnames =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">strand =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">gc =</span> <span class="kw">runif</span>(<span class="dv">7</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>()

grng</code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr1  [ 2,  1]      + |  0.76255108229816
##   [2]     chr1  [ 1,  1]      - | 0.669021712383255
##   [3]     chr2  [ 0,  1]      + | 0.204612161964178
##   [4]     chr1  [-1,  1]      - | 0.357524853432551
##   [5]     chr1  [13, 14]      - | 0.359475114848465
##   [6]     chr1  [14, 14]      + | 0.690290528349578
##   [7]     chr2  [15, 14]      + | 0.535811153938994
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="arithmetic-on-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#arithmetic-on-ranges" class="anchor"></a>Arithmetic on Ranges</h1>
<p>Sometimes you want to modify a genomic interval by altering the width of the interval while leaving the start, end or midpoint of the coordinates unaltered. This is achieved with the <code>set_width</code> verb along with <code>anchor_*</code> adverbs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>(<span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">8</span>)))
grng &lt;-<span class="st"> </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>(<span class="kw">data.frame</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), <span class="dt">end=</span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">8</span>), 
                          <span class="dt">seqnames =</span> <span class="st">"seq1"</span>,
                          <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"-"</span>)))
<span class="kw"><a href="../reference/set_width.html">set_width</a></span>(rng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_start</a></span>(rng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_end</a></span>(rng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4         5        10
##   [2]        -7         2        10
##   [3]        -1         8        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_center</a></span>(rng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -2         7        10
##   [2]        -3         6        10
##   [3]         1        10        10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_3p</a></span>(grng), <span class="dv">10</span>) <span class="co"># leave negative strand fixed</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1   [1, 10]      +
##   [2]     seq1   [2,  2]      *
##   [3]     seq1   [3,  8]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/set_width.html">set_width</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_5p</a></span>(grng), <span class="dv">10</span>) <span class="co"># leave positve strand fixed</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1   [ 1, 5]      +
##   [2]     seq1   [ 2, 2]      *
##   [3]     seq1   [-1, 8]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Similarly, you can modify the width of an interval using the <code>stretch</code> verb. Without anchoring, this function will extend the interval in either direction by an integer amount. With anchoring, either the start, end or midpoint are preserved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rng2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_center</a></span>(rng), <span class="dv">10</span>)
rng2</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -7        13        21
##   [2]        -8        12        21
##   [3]        -5        16        22</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_end</a></span>(rng2), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         3        13        11
##   [2]         2        12        11
##   [3]         5        16        12</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_start</a></span>(rng2), <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -7        23        31
##   [2]        -8        22        31
##   [3]        -5        26        32</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_3p</a></span>(grng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [-9, 15]      +
##   [2]     seq1  [ 2,  2]      *
##   [3]     seq1  [ 3,  8]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stretch.html">stretch</a></span>(<span class="kw"><a href="../reference/anchor.html">anchor_5p</a></span>(grng), <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [ 1,  5]      +
##   [2]     seq1  [ 2,  2]      *
##   [3]     seq1  [-7, 18]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>*Ranges</code> can be shifted left or right. If strand information is available we can also shift upstream or downstream.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_left</a></span>(rng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -9        -5         5
##   [2]        -8        -8         1
##   [3]        -7        -2         6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_right</a></span>(rng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        11        15         5
##   [2]        12        12         1
##   [3]        13        18         6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_upstream</a></span>(grng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [-9, -5]      +
##   [2]     seq1  [ 2,  2]      *
##   [3]     seq1  [13, 18]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/shift-ranges.html">shift_downstream</a></span>(grng, <span class="dv">10</span>)</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1  [11, 15]      +
##   [2]     seq1  [ 2,  2]      *
##   [3]     seq1  [-7, -2]      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="grouping-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#grouping-ranges" class="anchor"></a>Grouping Ranges</h1>
<p><code>plyranges</code> introduces a new class of <code>Ranges</code> called <code>RangesGrouped</code>, this is a similiar idea to the grouped <code>data.frame\tibble</code> in <code>dplyr</code>. Note that grouping does not change the structure of a <code>Ranges</code> object only how certain functions act on it.</p>
<p>Grouping can act on either the core components or the metadata columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grng &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">seqnames =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">strand =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">7</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
         <span class="dt">gc =</span> <span class="kw">runif</span>(<span class="dv">7</span>),
         <span class="dt">start =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">7</span>,
         <span class="dt">width =</span> <span class="dv">10</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>()

grng_by_strand &lt;-<span class="st"> </span>grng <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(strand)

grng_by_strand</code></pre></div>
<pre><code>## GRangesGrouped object with 7 ranges and 1 metadata column:
## Groups: strand
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2   [1, 10]      - | 0.889453538926318
##   [2]     chr2   [2, 11]      + | 0.180407245177776
##   [3]     chr2   [3, 12]      + | 0.629390850430354
##   [4]     chr1   [4, 13]      - | 0.989564136601985
##   [5]     chr1   [5, 14]      + | 0.130288870073855
##   [6]     chr2   [6, 15]      - | 0.330660525709391
##   [7]     chr2   [7, 16]      - | 0.865120546659455
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="restricting-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#restricting-ranges" class="anchor"></a>Restricting Ranges</h1>
<p>The verb <code>filter</code> can be used to restrict rows in the Ranges. Note that grouping will cause the <code>filter</code> to act within each group of the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grng <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(gc <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.3</span>)</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2   [2, 11]      + | 0.180407245177776
##   [2]     chr1   [5, 14]      + | 0.130288870073855
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filtering by group </span>
grng_by_strand <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(gc <span class="op">==</span><span class="st"> </span><span class="kw">max</span>(gc)) </code></pre></div>
<pre><code>## GRangesGrouped object with 2 ranges and 1 metadata column:
## Groups: 
##       seqnames    ranges strand |                gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;numeric&gt;
##   [1]     chr2   [3, 12]      + | 0.629390850430354
##   [2]     chr1   [4, 13]      - | 0.989564136601985
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We also provide the convience methods <code>filter_by_overlaps</code> and <code>filter_by_non_overlaps</code> for restricting by overlapping ranges.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>, <span class="dv">15</span>,<span class="dv">20</span>), <span class="dt">width =</span> <span class="dv">5</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>()
ir1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">start =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">6</span>, <span class="dt">width =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">7</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/Ranges.html">Ranges</a></span>()
ir0</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5
##   [3]        15        19         5
##   [4]        20        24         5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir1</code></pre></div>
<pre><code>## IRanges object with 5 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         4         3
##   [2]         3         6         4
##   [3]         4         8         5
##   [4]         5        10         6
##   [5]         6        12         7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter-overlaps.rd.html">filter_by_overlaps</a></span>(ir1)</code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter-overlaps.rd.html">filter_by_non_overlaps</a></span>(ir1) </code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        15        19         5
##   [2]        20        24         5</code></pre>
</div>
<div id="joins-or-another-way-at-looking-at-overlaps-between-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#joins-or-another-way-at-looking-at-overlaps-between-ranges" class="anchor"></a>Joins, or another way at looking at overlaps between Ranges</h1>
<p>We can think about finding overlaps as a type of join between two <code>*Ranges</code> where we are using the intervals as keys. Of most use are the <code>join_overlap_inner</code> and <code>join_overlap_left</code> functions. The former will return a <code>Ranges</code> object where the start, end, and width coordinates correspond to the amount of any overlap. By default, the width of the input ranges is added as metadata. This allows downstreaming filtering by the amount of overlap.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner</a></span>(ir0, ir1)</code></pre></div>
<pre><code>## IRanges object with 6 ranges and 2 metadata columns:
##           start       end     width |   width.x   width.y
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;integer&gt; &lt;integer&gt;
##   [1]         5         6         2 |         5         4
##   [2]         5         8         4 |         5         5
##   [3]         5         9         5 |         5         6
##   [4]         6         9         4 |         5         7
##   [5]        10        10         1 |         5         6
##   [6]        10        12         3 |         5         7</code></pre>
<p>The <code>join_overlap_left</code> method will add the coordinates of the right hand ranges that overlap those ranges on the left. Compared with <code>filter_by_overlaps</code> above, the overlap join expands the ranges to give information about each range on the right hand side that overlaps those on the left. We also provide a convienence method called <code>find_overlaps</code> that computes the same result.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_left</a></span>(ir0, ir1)</code></pre></div>
<pre><code>## IRanges object with 6 ranges and 3 metadata columns:
##           start       end     width |   start.y     end.y   width.y
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5 |         3         6         4
##   [2]         5         9         5 |         4         8         5
##   [3]         5         9         5 |         5        10         6
##   [4]         5         9         5 |         6        12         7
##   [5]        10        14         5 |         5        10         6
##   [6]        10        14         5 |         6        12         7</code></pre>
<p>We also provide methods for finding, nearest, preceding or following ranges. These methods nest a ranges column upon returning (maybe better to return just coordinates?).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/nearest-ranges.html">join_nearest</a></span>(ir0, ir1)</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |   nearest
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;IRanges&gt;
##   [1]         5         9         5 |   [6, 12]
##   [2]        10        14         5 |   [6, 12]
##   [3]        15        19         5 |   [6, 12]
##   [4]        20        24         5 |   [6, 12]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/follow-ranges.html">join_follow</a></span>(ir0, ir1)</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |   follows
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;IRanges&gt;
##   [1]         5         9         5 |   [2,  4]
##   [2]        10        14         5 |   [4,  8]
##   [3]        15        19         5 |   [6, 12]
##   [4]        20        24         5 |   [6, 12]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/precede-ranges.html">join_precede</a></span>(ir0, ir1) <span class="co"># nothing precedes returns empty ranges</span></code></pre></div>
<pre><code>## IRanges object with 0 ranges and 1 metadata column:
##        start       end     width |  precedes
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;IRanges&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/precede-ranges.html">join_precede</a></span>(ir1, ir0)</code></pre></div>
<pre><code>## IRanges object with 5 ranges and 1 metadata column:
##           start       end     width |  precedes
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;IRanges&gt;
##   [1]         2         4         3 |  [ 5,  9]
##   [2]         3         6         4 |  [10, 14]
##   [3]         4         8         5 |  [10, 14]
##   [4]         5        10         6 |  [15, 19]
##   [5]         6        12         7 |  [15, 19]</code></pre>
<p>For <code>GRanges</code> objects by default strand is not considered when performing overlap joins. To include strand when finding overlaps use append the prefix <code>directed</code>. To restrict overlapping ranges to those within the query range use the prefix <code>within</code>.</p>
<p>Itâ€™s also possible to group by overlaps. Using this approach we can count the number of overlaps that are greater than 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grp_by_olap &lt;-<span class="st"> </span>ir0 <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-overlaps.Rd.html">group_by_overlaps</a></span>(ir1)
grp_by_olap</code></pre></div>
<pre><code>## IRangesGrouped object with 6 ranges and 4 metadata columns:
## Groups: query
##           start       end     width | start.subject end.subject
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;integer&gt;   &lt;integer&gt;
##   [1]         5         9         5 |             3           6
##   [2]         5         9         5 |             4           8
##   [3]         5         9         5 |             5          10
##   [4]         5         9         5 |             6          12
##   [5]        10        14         5 |             5          10
##   [6]        10        14         5 |             6          12
##       width.subject     query
##           &lt;integer&gt; &lt;integer&gt;
##   [1]             4         1
##   [2]             5         1
##   [3]             6         1
##   [4]             7         1
##   [5]             6         2
##   [6]             7         2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grp_by_olap <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">n_overlaps =</span> <span class="kw">length</span>(query))</code></pre></div>
<pre><code>## IRangesGrouped object with 6 ranges and 5 metadata columns:
## Groups: query
##           start       end     width | start.subject end.subject
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |     &lt;integer&gt;   &lt;integer&gt;
##   [1]         5         9         5 |             3           6
##   [2]         5         9         5 |             4           8
##   [3]         5         9         5 |             5          10
##   [4]         5         9         5 |             6          12
##   [5]        10        14         5 |             5          10
##   [6]        10        14         5 |             6          12
##       width.subject     query n_overlaps
##           &lt;integer&gt; &lt;integer&gt;  &lt;integer&gt;
##   [1]             4         1          4
##   [2]             5         1          4
##   [3]             6         1          4
##   [4]             7         1          4
##   [5]             6         2          2
##   [6]             7         2          2</code></pre>
<p>Of course we can also add overlap counts via the <code>count_overlaps</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir0 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">n_overlaps =</span> <span class="kw"><a href="../reference/ranges-count-overlaps.Rd.html">count_overlaps</a></span>(., ir1))</code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width | n_overlaps
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |  &lt;integer&gt;
##   [1]         5         9         5 |          4
##   [2]        10        14         5 |          2
##   [3]        15        19         5 |          0
##   [4]        20        24         5 |          0</code></pre>
</div>
<div id="summarising-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#summarising-ranges" class="anchor"></a>Summarising Ranges</h1>
<p>The <code>summarise</code> function will return a tibble because the information required to return a _*Ranges_ object is lost. It is often most useful to use <code>summarise</code> in combination with the <code>group_by</code> family of functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ir1 &lt;-<span class="st"> </span>ir1 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">gc =</span> <span class="kw">runif</span>(<span class="kw">length</span>(.)))

ir0 <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-overlaps.Rd.html">group_by_overlaps</a></span>(ir1) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="dt">gc =</span> <span class="kw">mean</span>(gc))</code></pre></div>
<pre><code>## # A tibble: 2 x 2
##   query        gc
##   &lt;int&gt;     &lt;dbl&gt;
## 1     1 0.6755545
## 2     2 0.6357952</code></pre>
</div>
<div id="data-importoutput" class="section level1">
<h1 class="hasAnchor">
<a href="#data-importoutput" class="anchor"></a>Data Import/Output</h1>
<p>We provide convienence functions via <code>rtracklayer</code> for reading/writing the following data formats from/to _*Ranges_ objects.</p>
<ul>
<li>BED: <code>read/write_bed</code>
</li>
<li>BEDGraph: <code>read/write_bedgraph</code>
</li>
<li>GFF(1-3): <code>read/write_gff(1-3)</code>
</li>
<li>BigWig: <code>read/write_bw</code>
</li>
<li>Wig: <code>read/write_wig</code>
</li>
<li>narrowPeaks: <code>read/write_narrowpeaks</code>
</li>
</ul>
</div>
<div id="mapping-to-genomicrangesiranges" class="section level1">
<h1 class="hasAnchor">
<a href="#mapping-to-genomicrangesiranges" class="anchor"></a>Mapping to GenomicRanges/IRanges</h1>
<p>For users already familiar with the <code>IRanges</code> and <code>GenomicRanges</code> we provide mappings to <code>plyranges</code> functions.</p>
<div id="operations-on-range-width" class="section level2">
<h2 class="hasAnchor">
<a href="#operations-on-range-width" class="anchor"></a>Operations on range width</h2>
<p>For <code>GRanges</code> objects all functions ignore any strandedness, unless the strand of the range is anchored.</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="19%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>anchor_(start/end/center/3p/5p)</code></td>
<td>Fix the <code>start/end/center/</code> coordinates or positive/negative strand of range. Can be used in combination with any of the following</td>
<td>Available in functions that have a <code>fix</code> argument.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/set_width.html">set_width(x, width)</a></code></td>
<td>Modify the width of a <code>Range</code>
</td>
<td><code>resize</code></td>
</tr>
<tr class="odd">
<td><code><a href="../reference/stretch.html">stretch(x, extend)</a></code></td>
<td>Extend the start and end coordinates in opposite directions by a fixed amount.</td>
<td>
<code>start(x)&lt;- start(x) + extend%/%2</code>, <code>end(x) &lt;- end(x) -extend%/%2</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="operations-on-range-width-invariant" class="section level2">
<h2 class="hasAnchor">
<a href="#operations-on-range-width-invariant" class="anchor"></a>Operations on range width (invariant)</h2>
<table class="table">
<colgroup>
<col width="33%">
<col width="19%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>shift_[left/right/downstream/upstream](x, shift)</code></td>
<td>Shift the coordinates of the interval (left/right/downstream/upstream) by an integer amount.</td>
<td>
<code>shift_right</code> corresponds to <code>shift</code>
</td>
</tr>
<tr class="even">
<td><code>flank_[left/right/downstream/upstream](x, width)</code></td>
<td>Generates flanking regions of size width <code>left/right/downstream/upstream/</code>
</td>
<td>corresponds to <code>flank</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="set-operations-vector-wise" class="section level2">
<h2 class="hasAnchor">
<a href="#set-operations-vector-wise" class="anchor"></a>Set operations (vector wise)</h2>
<p>These are usual set-operations that act on the sets of the ranges represented in x and y. By default these operations will ignore any strand information. The directed versions of these functions will take into account strand.</p>
<table class="table">
<colgroup>
<col width="33%">
<col width="19%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>[intersect/setdiff/union/]_ranges</code></td>
<td>Set operations between two ranges, ignoring strand.</td>
<td>
<code>intersect/setdiff/union/</code> with <code>ignore.strand = FALSE</code>
</td>
</tr>
<tr class="even">
<td><code>[intersect/setdiff/union/]_anchored_ranges</code></td>
<td>As above taking into account strandedness.</td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="set-operations-element-wise" class="section level2">
<h2 class="hasAnchor">
<a href="#set-operations-element-wise" class="anchor"></a>Set operations (element wise)</h2>
<p>We provide infix operators and the verbs between and span to the represent element wise range operations. These map to the <code>pintersect/punion/psetdiff/pgap/punion(fill.gap = FALSE)</code> functions.</p>
</div>
<div id="restrictions" class="section level2">
<h2 class="hasAnchor">
<a href="#restrictions" class="anchor"></a>Restrictions</h2>
<p>The verb <code>filter</code> corresponds to <code>subset</code>, while <code>filter_by_[overlaps/non_overlaps]</code> corresponds to <code>subsetByOverlaps</code>.</p>
</div>
<div id="aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregation" class="anchor"></a>Aggregation</h2>
<p>The <code>summarise</code> verb is most similar to the <code>aggregate</code> methods defined in <code>GenomicRanges/IRanges</code>.</p>
<p>The <code>reduce_ranges/disjoin_ranges</code> correspond to the <code>reduce/disjoin</code> methods. However, the former methods allow additional summarisation.</p>
<p>The <code><a href="../reference/set_coverage.html">set_coverage(x)</a></code> method corresponds to <code>[I/G]ranges(coverage(x))</code>.</p>
</div>
<div id="overlaps" class="section level2">
<h2 class="hasAnchor">
<a href="#overlaps" class="anchor"></a>Overlaps</h2>
<p>For <code>GRanges</code> objects all functions ignore any strandedness, unless the suffix <code>directed</code> is added to the funciton call</p>
<table style="width:100%;" class="table">
<colgroup>
<col width="25%">
<col width="54%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th>
<code>plyranges</code> function</th>
<th>Description</th>
<th>
<code>GenomicRanges/IRanges</code> command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/ranges-overlaps.Rd.html">find_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns a <code>Ranges</code> object with any range in <code>y</code> that overlaps <code>x</code>. Appends the metadata in <code>y</code> and its genomic intervals to the returning <code>Ranges</code>.</td>
<td>
<code>findOverlaps(x,y, maxgap, minoverlap, type = "any")</code> with expanding <code>x</code> and <code>y</code> by their hits and appending the <code>mcols</code> in <code>y</code>.</td>
</tr>
<tr class="even">
<td><code><a href="../reference/ranges-overlaps.Rd.html">group_by_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns a GroupedRanges object grouped by the query hits.</td>
<td>Same as above with an additional column called <code>query</code> which contains the queryHits.</td>
</tr>
<tr class="odd">
<td><code><a href="../reference/ranges-count-overlaps.Rd.html">count_overlaps(x, y, maxgap, minoverlap)</a></code></td>
<td>Returns an integer vector (used with <code>mutate</code>)</td>
<td><code>countOverlaps(x, y, maxgap, minoverlap, type = "any")</code></td>
</tr>
<tr class="even">
<td><code><a href="../reference/self-overlap-joins.html">join_overlap_self(x, maxgap, minoverlap)</a></code></td>
<td>Returns a <code>Ranges</code> object with any range that overlaps itself.</td>
<td><code>findOverlaps(x,x, maxgap, minoverlap, type = "any")</code></td>
</tr>
<tr class="odd">
<td><code><a href="../reference/overlap-joins.html">join_overlap_inner(x, y, maxgap, minoverlap)</a></code></td>
<td>Finds the intersecting ranges that overlap in <code>x</code> and <code>y</code>. Returns a <code>Ranges</code> object with the metadata from <code>x</code> and <code>y</code>.</td>
<td>
<code>findOverlapsPairs(x,y, maxgap, minoverlap, type = "any")</code> + <code>pintersect</code>.</td>
</tr>
<tr class="even">
<td><code>*_within</code></td>
<td>Adding suffix <code>within</code> will find overlaps</td>
<td>Makes <code>type = "within"</code>
</td>
</tr>
<tr class="odd">
<td><code>*_includes</code></td>
<td>inverse of within functions</td>
<td>-</td>
</tr>
<tr class="even">
<td><code>join_nearest[_left/right/up/downstream](x,y)</code></td>
<td>Finds nearest neighbour ranges between <code>x</code> and <code>y</code>.</td>
<td>
<code>nearest</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
<tr class="odd">
<td><code>join_precede[_left/right/up/downstream](x,y)</code></td>
<td>Finds ranges in <code>x</code> that preced <code>y</code>
</td>
<td>
<code>precedes</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
<tr class="even">
<td><code>join_follow[_left/right/up/downstream](x,y)</code></td>
<td>Finds ranges in <code>x</code> that follow <code>y</code>
</td>
<td>
<code>precedes</code> + reindexing to return a <code>Ranges</code> object.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="appendix" class="section level1">
<h1 class="hasAnchor">
<a href="#appendix" class="anchor"></a>Appendix</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.2 (2017-09-28)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] plyranges_0.1.0      GenomicRanges_1.30.0 GenomeInfoDb_1.14.0 
## [4] IRanges_2.12.0       S4Vectors_0.16.0     BiocGenerics_0.24.0 
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.13               compiler_3.4.2            
##  [3] XVector_0.18.0             bindr_0.1                 
##  [5] bitops_1.0-6               tools_3.4.2               
##  [7] zlibbioc_1.24.0            digest_0.6.12             
##  [9] tibble_1.3.4               evaluate_0.10.1           
## [11] lattice_0.20-35            pkgconfig_2.0.1           
## [13] rlang_0.1.4.9000           Matrix_1.2-11             
## [15] DelayedArray_0.4.1         yaml_2.1.14               
## [17] bindrcpp_0.2               GenomeInfoDbData_0.99.1   
## [19] rtracklayer_1.38.0         stringr_1.2.0             
## [21] dplyr_0.7.4                knitr_1.17                
## [23] Biostrings_2.46.0          tidyselect_0.2.3          
## [25] rprojroot_1.2              grid_3.4.2                
## [27] glue_1.2.0                 Biobase_2.38.0            
## [29] R6_2.2.2                   XML_3.98-1.9              
## [31] BiocParallel_1.12.0        rmarkdown_1.8             
## [33] purrr_0.2.4                tidyr_0.7.2               
## [35] magrittr_1.5               backports_1.1.1           
## [37] Rsamtools_1.30.0           htmltools_0.3.6           
## [39] matrixStats_0.52.2         GenomicAlignments_1.14.0  
## [41] assertthat_0.2.0           SummarizedExperiment_1.8.0
## [43] stringi_1.1.5              RCurl_1.95-4.8</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#ranges-revisted">Ranges revisted</a></li>
      <li><a href="#constructing-ranges">Constructing Ranges</a></li>
      <li><a href="#arithmetic-on-ranges">Arithmetic on Ranges</a></li>
      <li><a href="#grouping-ranges">Grouping Ranges</a></li>
      <li><a href="#restricting-ranges">Restricting Ranges</a></li>
      <li><a href="#joins-or-another-way-at-looking-at-overlaps-between-ranges">Joins, or another way at looking at overlaps between Ranges</a></li>
      <li><a href="#summarising-ranges">Summarising Ranges</a></li>
      <li><a href="#data-importoutput">Data Import/Output</a></li>
      <li>
<a href="#mapping-to-genomicrangesiranges">Mapping to GenomicRanges/IRanges</a><ul class="nav nav-pills nav-stacked">
<li><a href="#operations-on-range-width">Operations on range width</a></li>
      <li><a href="#operations-on-range-width-invariant">Operations on range width (invariant)</a></li>
      <li><a href="#set-operations-vector-wise">Set operations (vector wise)</a></li>
      <li><a href="#set-operations-element-wise">Set operations (element wise)</a></li>
      <li><a href="#restrictions">Restrictions</a></li>
      <li><a href="#aggregation">Aggregation</a></li>
      <li><a href="#overlaps">Overlaps</a></li>
      </ul>
</li>
      <li><a href="#appendix">Appendix</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Stuart Lee, Michael Lawrence, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
