<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Common Workflow Solutions â€¢ plyranges</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Common Workflow Solutions">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plyranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/an-introduction.html">Introduction to plyranges</a>
    </li>
    <li>
      <a href="../articles/common-tasks.html">Common genomic data wrangling tasks</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/plyranges">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Common Workflow Solutions</h1>
                        <h4 class="author">Stuart Lee</h4>
            
            <h4 class="date">27/11/2018</h4>
      
      
      <div class="hidden name"><code>solutions.Rmd</code></div>

    </div>

    
    
<p>These are the solutions to the excercies found in <a href="common-tasks.html">Common Genomic Data Wrangling Tasks</a>.</p>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>We need to read in three files from HelloRangesData</p>
<ul>
<li>exons.bed (RefSeq hg19 annotated exons)</li>
<li>cpg.bed (RefSeq hg19 annotated CpG islands)</li>
<li>hesc.chromHmm.bed (predicted functional elements from chromHMM program)</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(plyranges))
<span class="co"># our genome build</span>
build &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-info.html">genome_info</a></span>(<span class="st">"hg19"</span>)

exons &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"exons.bed"</span>, <span class="dt">package=</span><span class="st">"HelloRangesData"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/io-bed-read.html">read_bed</a></span>(<span class="dt">genome_info =</span> build)

cpg &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"cpg.bed"</span>, <span class="dt">package=</span><span class="st">"HelloRangesData"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/io-bed-read.html">read_bed</a></span>(<span class="dt">genome_info =</span> build)

hesc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"hesc.chromHmm.bed"</span>, <span class="dt">package=</span><span class="st">"HelloRangesData"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/io-bed-read.html">read_bed</a></span>(<span class="dt">genome_info =</span> build)</code></pre></div>
<div id="solutions" class="section level2">
<h2 class="hasAnchor">
<a href="#solutions" class="anchor"></a>Solutions</h2>
<div id="filtering-and-mutating" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-and-mutating" class="anchor"></a>Filtering and mutating</h3>
<ol style="list-style-type: decimal">
<li>Modify the above example so only CpG islands that are completely within exons are included in the result.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">is_contained =</span> . <span class="op">%within%</span><span class="st"> </span>exons) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(is_contained)</code></pre></div>
<pre><code>## GRanges object with 2235 ranges and 2 metadata columns:
##          seqnames            ranges strand |        name is_contained
##             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;logical&gt;
##      [1]     chr1     135125-135563      * |     CpG:_30         TRUE
##      [2]     chr1     327791-328229      * |     CpG:_29         TRUE
##      [3]     chr1     788864-789211      * |     CpG:_28         TRUE
##      [4]     chr1     990275-990694      * |     CpG:_29         TRUE
##      [5]     chr1   1132828-1133218      * |     CpG:_40         TRUE
##      ...      ...               ...    ... .         ...          ...
##   [2231]     chrY     171535-171828      * |     CpG:_23         TRUE
##   [2232]     chrY   1496664-1497002      * |     CpG:_32         TRUE
##   [2233]     chrY   2356771-2358703      * |    CpG:_133         TRUE
##   [2234]     chrY 15815489-15815779      * |     CpG:_26         TRUE
##   [2235]     chrY 16941823-16942188      * |     CpG:_32         TRUE
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or just </span>
<span class="co"># cpg %&gt;% filter(. %within% exons)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Create a new column in exons called tx_id (the transcript id is everything before _exon in name).</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we can use the sub function in base R</span>
exons &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">tx_id =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">sub</a></span>(<span class="st">"_exon.*"</span>, <span class="st">""</span>, name))</code></pre></div>
</div>
<div id="modifying-genomic-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#modifying-genomic-regions" class="anchor"></a>Modifying genomic regions</h3>
<ol style="list-style-type: decimal">
<li>Create a new GRanges object from CpG islands that stretches the intervals on either side by their width while leaving the centre of the interval fixed.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpg_stretch &lt;-<span class="st"> </span>cpg <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-anchor.html">anchor_centre</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">width =</span> <span class="dv">2</span><span class="op">*</span>width)

<span class="co"># alternative is to use `stretch()`</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Create a new GRanges object from exons that has only non-exonic regions.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">no_exon &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/ranges-setops.html">complement_ranges</a></span>()</code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Create a new GRanges object that represent 2bp canonical splice sites on either side of exon.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># you can do this directly with mutate and anchors</span>
left &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-anchor.html">anchor_start</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">start =</span> start <span class="op">-</span><span class="st"> </span>2L, <span class="dt">width =</span> 2L)

right &lt;-<span class="st"> </span>exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-anchor.html">anchor_end</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">end =</span> end <span class="op">+</span><span class="st"> </span>2L,
         <span class="dt">width =</span> 2L)

sites &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ranges-bind.html">bind_ranges</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">left =</span> left, <span class="dt">right =</span> right), <span class="dt">.id =</span> <span class="st">"side"</span>)

<span class="co"># or with flank_* family of functions</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/identical">identical</a></span>(exons <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/flank-ranges.html">flank_left</a></span>(<span class="dv">2</span>), left)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/identical">identical</a></span>(exons <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/flank-ranges.html">flank_right</a></span>(<span class="dv">2</span>), right)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="summarising-granges-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#summarising-granges-objects" class="anchor"></a>Summarising GRanges objects</h3>
<ol style="list-style-type: decimal">
<li>How many ranges are there for each predicted state?</li>
<li>How many base pairs are represented in the genome for each predicted state? Which state has the maximum number of bases?</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">state_summary &lt;-<span class="st"> </span>hesc <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(name) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(
    <span class="dt">n_ranges =</span> <span class="kw"><a href="../reference/n.html">n</a></span>(), 
    <span class="dt">n_bases =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(width)
  )

state_summary <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(n_bases <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Extremes">max</a></span>(n_bases))</code></pre></div>
<pre><code>##                name n_ranges    n_bases
## 1 13_Heterochrom/lo    91228 1992618522</code></pre>
</div>
<div id="overlaps" class="section level3">
<h3 class="hasAnchor">
<a href="#overlaps" class="anchor"></a>Overlaps</h3>
<ol style="list-style-type: decimal">
<li>Create a new GRanges object, that has exons that are completely within Enhancers elements of hesc. How many exons are there?</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enhancers &lt;-<span class="st"> </span>hesc <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grepl</a></span>(<span class="st">"Enhancer"</span>, name))

exon_within &lt;-<span class="st"> </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_inner_within</a></span>(exons, enhancers)

exon_within</code></pre></div>
<pre><code>## GRanges object with 13746 ranges and 4 metadata columns:
##           seqnames              ranges strand |
##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##       [1]     chr1       948847-948956      + |
##       [2]     chr1     1051440-1051736      - |
##       [3]     chr1     1109286-1109306      + |
##       [4]     chr1     1109804-1109869      + |
##       [5]     chr1     1219358-1219470      + |
##       ...      ...                 ...    ... .
##   [13742]     chrX 153927569-153927788      - |
##   [13743]     chrX 153927569-153927788      - |
##   [13744]     chrX 153927569-153927788      - |
##   [13745]     chrX 153927569-153927788      - |
##   [13746]     chrX 154114409-154114577      - |
##                                           name.x     score        tx_id
##                                      &lt;character&gt; &lt;numeric&gt;  &lt;character&gt;
##       [1]       NM_005101_exon_0_0_chr1_948847_f         0    NM_005101
##       [2]      NM_017891_exon_9_0_chr1_1051440_r         0    NM_017891
##       [3]   NM_001130045_exon_0_0_chr1_1109286_f         0 NM_001130045
##       [4]   NM_001130045_exon_2_0_chr1_1109804_f         0 NM_001130045
##       [5]   NM_001130413_exon_4_0_chr1_1219358_f         0 NM_001130413
##       ...                                    ...       ...          ...
##   [13742] NM_001081573_exon_4_0_chrX_153927569_r         0 NM_001081573
##   [13743] NM_001282283_exon_3_0_chrX_153927569_r         0 NM_001282283
##   [13744]    NM_080612_exon_4_0_chrX_153927569_r         0    NM_080612
##   [13745]    NR_104114_exon_4_0_chrX_153927569_r         0    NR_104114
##   [13746]    NM_019863_exon_4_0_chrX_154114409_r         0    NM_019863
##                      name.y
##                 &lt;character&gt;
##       [1] 4_Strong_Enhancer
##       [2]   6_Weak_Enhancer
##       [3]   6_Weak_Enhancer
##       [4]   6_Weak_Enhancer
##       [5]   7_Weak_Enhancer
##       ...               ...
##   [13742]   6_Weak_Enhancer
##   [13743]   6_Weak_Enhancer
##   [13744]   6_Weak_Enhancer
##   [13745]   6_Weak_Enhancer
##   [13746]   6_Weak_Enhancer
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Use <code><a href="../reference/overlap-joins.html">join_overlap_intersect()</a></code> to filter exons if at least 50 per cent of their bases are overlapped by enhancer elements.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">mutate</a></span>(<span class="dt">total_length =</span> width) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_intersect</a></span>(enhancers) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">filter</a></span>(width <span class="op">/</span><span class="st"> </span>total_length <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>)</code></pre></div>
<pre><code>## GRanges object with 21151 ranges and 5 metadata columns:
##           seqnames              ranges strand |
##              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |
##       [1]     chr1         35738-35937      - |
##       [2]     chr1         35738-35937      - |
##       [3]     chr1       910938-911537      - |
##       [4]     chr1       911938-912004      - |
##       [5]     chr1       948847-948956      + |
##       ...      ...                 ...    ... .
##   [21147]     chrX 153927569-153927788      - |
##   [21148]     chrX 154114409-154114577      - |
##   [21149]     chrX 154300607-154300618      + |
##   [21150]     chrX 154300607-154300618      + |
##   [21151]     chrX 154300607-154300618      + |
##                                           name.x     score        tx_id
##                                      &lt;character&gt; &lt;numeric&gt;  &lt;character&gt;
##       [1]        NR_026818_exon_2_0_chr1_35721_r         0    NR_026818
##       [2]        NR_026820_exon_2_0_chr1_35721_r         0    NR_026820
##       [3]       NR_027693_exon_0_0_chr1_910579_r         0    NR_027693
##       [4]       NR_027693_exon_1_0_chr1_911879_r         0    NR_027693
##       [5]       NM_005101_exon_0_0_chr1_948847_f         0    NM_005101
##       ...                                    ...       ...          ...
##   [21147]    NR_104114_exon_4_0_chrX_153927569_r         0    NR_104114
##   [21148]    NM_019863_exon_4_0_chrX_154114409_r         0    NM_019863
##   [21149] NM_001018055_exon_1_0_chrX_154300602_f         0 NM_001018055
##   [21150] NM_001242640_exon_1_0_chrX_154300602_f         0 NM_001242640
##   [21151]    NM_024332_exon_1_0_chrX_154300602_f         0    NM_024332
##           total_length            name.y
##              &lt;integer&gt;       &lt;character&gt;
##       [1]          361   7_Weak_Enhancer
##       [2]          361   7_Weak_Enhancer
##       [3]         1071   6_Weak_Enhancer
##       [4]          126   6_Weak_Enhancer
##       [5]          110 4_Strong_Enhancer
##       ...          ...               ...
##   [21147]          220   6_Weak_Enhancer
##   [21148]          169   6_Weak_Enhancer
##   [21149]           17   6_Weak_Enhancer
##   [21150]           17   6_Weak_Enhancer
##   [21151]           17   6_Weak_Enhancer
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Count the number of each enchancer element type that are exonic. There are several ways of doing this, but see if you can come up with a solution using <code><a href="../reference/overlap-joins.html">join_overlap_left()</a></code> + <code><a href="../reference/ranges-disjoin.html">disjoin_ranges()</a></code>.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enhancers <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/overlap-joins.html">join_overlap_left</a></span>(exons) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(name.x) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/ranges-disjoin.html">disjoin_ranges</a></span>(<span class="dt">olap_any =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all">all</a></span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(tx_id))) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">group_by</a></span>(name.x) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidyverse-reexports.html">summarise</a></span>(<span class="dt">prop =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(olap_any) <span class="op">/</span><span class="st"> </span><span class="kw"><a href="../reference/n.html">n</a></span>())</code></pre></div>
<pre><code>## DataFrame with 4 rows and 2 columns
##              name.x               prop
##         &lt;character&gt;          &lt;numeric&gt;
## 1 4_Strong_Enhancer  0.129369300911854
## 2 5_Strong_Enhancer 0.0830471433031533
## 3   6_Weak_Enhancer 0.0921656261823685
## 4   7_Weak_Enhancer 0.0614762445084519</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#data">Data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#solutions">Solutions</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://stuartlee.org">Stuart Lee</a>, <a href="https://github.com/lawremi">Michael Lawrence</a>, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
